{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a2b8b81d-48be-489b-a28f-3856205f52f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy as sp \n",
    "import numpy as np\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "313f2c5f-27a9-4c03-a99f-37e559e405dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain Identifier</th>\n",
       "      <th>Strain name</th>\n",
       "      <th>Evolution Experiment?</th>\n",
       "      <th>Parental Strain</th>\n",
       "      <th>Alternative Strain Name</th>\n",
       "      <th>Full Strain Name After Genome Analysis</th>\n",
       "      <th>ENA Accession</th>\n",
       "      <th>Alternative Genome Source</th>\n",
       "      <th>Reference (PMID)</th>\n",
       "      <th>Isolation origin</th>\n",
       "      <th>Year</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>German BSL level</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NT12447</td>\n",
       "      <td>B367</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KC329</td>\n",
       "      <td>Escherichia coli B367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Broad institute Antibiotic resistance collection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Australian magpie Coffs</td>\n",
       "      <td>2001</td>\n",
       "      <td>Commensal strain</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NT12619</td>\n",
       "      <td>S17</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli S17</td>\n",
       "      <td>PRJNA189395</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23516222</td>\n",
       "      <td>Broiler chick with septicemia</td>\n",
       "      <td>2013</td>\n",
       "      <td>Pathogenic strain</td>\n",
       "      <td>2.0</td>\n",
       "      <td>APEC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Strain Identifier Strain name  Evolution Experiment? Parental Strain  \\\n",
       "0           NT12447        B367                      0             NaN   \n",
       "1           NT12619         S17                      0             NaN   \n",
       "\n",
       "  Alternative Strain Name Full Strain Name After Genome Analysis  \\\n",
       "0                   KC329                  Escherichia coli B367   \n",
       "1                     NaN                   Escherichia coli S17   \n",
       "\n",
       "  ENA Accession                         Alternative Genome Source  \\\n",
       "0           NaN  Broad institute Antibiotic resistance collection   \n",
       "1   PRJNA189395                                               NaN   \n",
       "\n",
       "  Reference (PMID)               Isolation origin  Year          Phenotype  \\\n",
       "0              NaN       Australian magpie Coffs   2001   Commensal strain   \n",
       "1         23516222  Broiler chick with septicemia  2013  Pathogenic strain   \n",
       "\n",
       "   German BSL level Notes  \n",
       "0               1.0   NaN  \n",
       "1               2.0  APEC  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strains = pd.read_csv('/Users/bamudamaris/Library/CloudStorage/GoogleDrive-fufordamaris@gmail.com/My Drive/Data/Datasets/mystrains.tsv', sep = '\\t')\n",
    "strains.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bde5a177-0764-47b4-8b8a-72a4df1e717a",
   "metadata": {},
   "source": [
    "### Focus only genes with  patterns significantly associated with their expression. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e9229c5f-4ba5-4662-af6d-34377efcf8bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sig_patterns = [p.split('.tsv')[0] for p in os.listdir('/Users/bamudamaris/ownCloud2/E.coli/GWAS/2023_10_GWAS/panfeed/associations/sig/')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bf30f701-69c5-4d3f-a915-e01c6096c66f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene_names = [p.split('.tsv')[0] for p in os.listdir('/Users/bamudamaris/ownCloud2/E.coli/GWAS/2023_10_GWAS/panfeed/associations/sig/')]\n",
    "\n",
    "# Directory containing the files\n",
    "directory = '/Users/bamudamaris/ownCloud2/model/ecoli/tx-panfeed_230_rerun/tested_patterns/'\n",
    "\n",
    "# List to store the contents of each file\n",
    "all_contents = []\n",
    "\n",
    "# Loop through all files in the directory\n",
    "for filename in os.listdir(directory):\n",
    "    if filename.endswith('.tsv'):\n",
    "        name = filename.split('.tsv')[0]\n",
    "        with open(os.path.join(directory, filename), 'r') as file:\n",
    "            for line in file:\n",
    "                line = line.strip()  # Remove leading/trailing whitespace\n",
    "                all_contents.append(f\"{name}\\t{line}\")  # Add the filename as a new column\n",
    "\n",
    "# Create a DataFrame from the concatenated contents\n",
    "concatenated_df = pd.DataFrame([line.split('\\t') for line in all_contents])\n",
    "\n",
    "# Rename columns for better understanding\n",
    "concatenated_df.columns = ['gene_name', 'variant', 'ltr-pvalue', 'beta', 'k-samples']\n",
    "\n",
    "# Filter the DataFrame to keep only the rows where the gene name starts with any of the gene names from the gene_names DataFrame\n",
    "filtered_df = concatenated_df[concatenated_df['gene_name'].apply(lambda x: any(x.startswith(gene) for gene in gene_names))]\n",
    "\n",
    "# Save the filtered DataFrame to a new file\n",
    "filtered_df.to_csv('filtered_concatenated_patterns.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8b2ab1dc-d6c9-41ca-b2a9-c6c1618c3fdc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cds-NP_417082.1-613</th>\n",
       "      <th>cds-NP_417082.1-607</th>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <th>cds-NP_417420.1-2778</th>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <th>cds-NP_417420.1-2793</th>\n",
       "      <th>cds-NP_417420.1-2787</th>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <th>...</th>\n",
       "      <th>cds-NP_418440.1-1100</th>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <th>cds-NP_418618.1-1642</th>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <th>cds-NP_414578.2-87</th>\n",
       "      <th>cds-NP_414578.2-93</th>\n",
       "      <th>cds-NP_414578.2-78</th>\n",
       "      <th>cds-NP_417082.1-622</th>\n",
       "      <th>cds-NP_417082.1-636</th>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strain</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NT12001</th>\n",
       "      <td>13023.185487</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>2057.162663</td>\n",
       "      <td>1452.116028</td>\n",
       "      <td>4171.010919</td>\n",
       "      <td>2767.218986</td>\n",
       "      <td>1711.620913</td>\n",
       "      <td>2767.218986</td>\n",
       "      <td>2767.218986</td>\n",
       "      <td>11616.601043</td>\n",
       "      <td>...</td>\n",
       "      <td>1284.565436</td>\n",
       "      <td>2644.365886</td>\n",
       "      <td>2279.413599</td>\n",
       "      <td>2281.612175</td>\n",
       "      <td>2057.162663</td>\n",
       "      <td>2057.162663</td>\n",
       "      <td>2057.162663</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>2922.76381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12002</th>\n",
       "      <td>13023.185487</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>2057.162663</td>\n",
       "      <td>1452.116028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2767.218986</td>\n",
       "      <td>1711.620913</td>\n",
       "      <td>2767.218986</td>\n",
       "      <td>2767.218986</td>\n",
       "      <td>11616.601043</td>\n",
       "      <td>...</td>\n",
       "      <td>1284.565436</td>\n",
       "      <td>2644.365886</td>\n",
       "      <td>2279.413599</td>\n",
       "      <td>2281.612175</td>\n",
       "      <td>2057.162663</td>\n",
       "      <td>2057.162663</td>\n",
       "      <td>2057.162663</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>2922.76381</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 539 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         cds-NP_417082.1-613  cds-NP_417082.1-607  cds-NP_414578.2-75  \\\n",
       "strain                                                                  \n",
       "NT12001         13023.185487         13023.185487         2057.162663   \n",
       "NT12002         13023.185487         13023.185487         2057.162663   \n",
       "\n",
       "         cds-NP_418204.2-940  cds-NP_414880.2-2266  cds-NP_417420.1-2778  \\\n",
       "strain                                                                     \n",
       "NT12001          1452.116028           4171.010919           2767.218986   \n",
       "NT12002          1452.116028                   NaN           2767.218986   \n",
       "\n",
       "         cds-NP_418208.1-940  cds-NP_417420.1-2793  cds-NP_417420.1-2787  \\\n",
       "strain                                                                     \n",
       "NT12001          1711.620913           2767.218986           2767.218986   \n",
       "NT12002          1711.620913           2767.218986           2767.218986   \n",
       "\n",
       "         cds-NP_416773.1-1254  ...  cds-NP_418440.1-1100  \\\n",
       "strain                         ...                         \n",
       "NT12001          11616.601043  ...           1284.565436   \n",
       "NT12002          11616.601043  ...           1284.565436   \n",
       "\n",
       "         cds-NP_418554.1-1855  cds-NP_418618.1-1642  cds-NP_418686.1-718  \\\n",
       "strain                                                                     \n",
       "NT12001           2644.365886           2279.413599          2281.612175   \n",
       "NT12002           2644.365886           2279.413599          2281.612175   \n",
       "\n",
       "         cds-NP_414578.2-87  cds-NP_414578.2-93  cds-NP_414578.2-78  \\\n",
       "strain                                                                \n",
       "NT12001         2057.162663         2057.162663         2057.162663   \n",
       "NT12002         2057.162663         2057.162663         2057.162663   \n",
       "\n",
       "         cds-NP_417082.1-622  cds-NP_417082.1-636  cds-YP_588463.1-2320  \n",
       "strain                                                                   \n",
       "NT12001         13023.185487         13023.185487            2922.76381  \n",
       "NT12002         13023.185487         13023.185487            2922.76381  \n",
       "\n",
       "[2 rows x 539 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#file generated from the in_silco_validation_panfeed_rerun script. \n",
    "f_df = pd.read_csv('filtered_concatenated_patterns.tsv', sep = '\\t')\n",
    "tx_rates = pd.read_csv('tx_rates_panfeed_230.tsv', sep = '\\t', index_col = 0)\n",
    "\n",
    "# Filter tx_rates to only include columns that are in the index of f_df\n",
    "filtered_columns = tx_rates.columns.intersection(f_df['gene_name'])\n",
    "tx_rates_filtered = tx_rates[filtered_columns]\n",
    "tx_rates_filtered.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "55c247b5-35b0-48e6-8b5b-483b5865a0dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cds-NP_417082.1-613</th>\n",
       "      <th>cds-NP_417082.1-607</th>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <th>cds-NP_417420.1-2778</th>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <th>cds-NP_417420.1-2793</th>\n",
       "      <th>cds-NP_417420.1-2787</th>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <th>...</th>\n",
       "      <th>cds-NP_418440.1-1100</th>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <th>cds-NP_418618.1-1642</th>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <th>cds-NP_414578.2-87</th>\n",
       "      <th>cds-NP_414578.2-93</th>\n",
       "      <th>cds-NP_414578.2-78</th>\n",
       "      <th>cds-NP_417082.1-622</th>\n",
       "      <th>cds-NP_417082.1-636</th>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strain</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NT12001</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12002</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12004</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12012</th>\n",
       "      <td>-2.117602</td>\n",
       "      <td>-2.117602</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.318827</td>\n",
       "      <td>-1.464291</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.152345</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.960852</td>\n",
       "      <td>0.476826</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.117602</td>\n",
       "      <td>-2.117602</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 539 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         cds-NP_417082.1-613  cds-NP_417082.1-607  cds-NP_414578.2-75  \\\n",
       "strain                                                                  \n",
       "NT12001             0.000000             0.000000                 0.0   \n",
       "NT12002             0.000000             0.000000                 0.0   \n",
       "NT12004             0.000000             0.000000                 0.0   \n",
       "NT12012            -2.117602            -2.117602                 0.0   \n",
       "\n",
       "         cds-NP_418204.2-940  cds-NP_414880.2-2266  cds-NP_417420.1-2778  \\\n",
       "strain                                                                     \n",
       "NT12001             0.000000              0.000000              0.000000   \n",
       "NT12002             0.000000                   NaN              0.000000   \n",
       "NT12004             0.000000              0.000000              0.000000   \n",
       "NT12012             0.318827             -1.464291              0.000012   \n",
       "\n",
       "         cds-NP_418208.1-940  cds-NP_417420.1-2793  cds-NP_417420.1-2787  \\\n",
       "strain                                                                     \n",
       "NT12001                  0.0              0.000000              0.000000   \n",
       "NT12002                  0.0              0.000000              0.000000   \n",
       "NT12004                  0.0              0.000000              0.000000   \n",
       "NT12012                  0.0              0.000012              0.000012   \n",
       "\n",
       "         cds-NP_416773.1-1254  ...  cds-NP_418440.1-1100  \\\n",
       "strain                         ...                         \n",
       "NT12001                   0.0  ...              0.000000   \n",
       "NT12002                   0.0  ...              0.000000   \n",
       "NT12004                   0.0  ...              0.000000   \n",
       "NT12012                   0.0  ...              0.152345   \n",
       "\n",
       "         cds-NP_418554.1-1855  cds-NP_418618.1-1642  cds-NP_418686.1-718  \\\n",
       "strain                                                                     \n",
       "NT12001                   0.0              0.000000             0.000000   \n",
       "NT12002                   0.0              0.000000             0.000000   \n",
       "NT12004                   0.0              0.000000             0.000000   \n",
       "NT12012                   0.0             -0.960852             0.476826   \n",
       "\n",
       "         cds-NP_414578.2-87  cds-NP_414578.2-93  cds-NP_414578.2-78  \\\n",
       "strain                                                                \n",
       "NT12001                 0.0                 0.0                 0.0   \n",
       "NT12002                 0.0                 0.0                 0.0   \n",
       "NT12004                 0.0                 0.0                 0.0   \n",
       "NT12012                 0.0                 0.0                 0.0   \n",
       "\n",
       "         cds-NP_417082.1-622  cds-NP_417082.1-636  cds-YP_588463.1-2320  \n",
       "strain                                                                   \n",
       "NT12001             0.000000             0.000000                   0.0  \n",
       "NT12002             0.000000             0.000000                   0.0  \n",
       "NT12004             0.000000             0.000000                   0.0  \n",
       "NT12012            -2.117602            -2.117602                   NaN  \n",
       "\n",
       "[4 rows x 539 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#correlation between log2Foldchange and predicted transcription rates \n",
    "#tx_rates of each divided by that of k12 and then log2 transformed.\n",
    "df_new = np.log2(tx_rates_filtered/tx_rates_filtered.loc['NT12001'])\n",
    "df_new.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "43f629ee-4725-451c-88b8-103ef66d5888",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_new.to_csv('predicted_rates_sig_associations.tsv', sep ='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20337167-11f6-44bc-a441-3b3693a91999",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene-id</th>\n",
       "      <th>cds-NP_414543.1</th>\n",
       "      <th>cds-NP_414544.1</th>\n",
       "      <th>cds-NP_414545.1</th>\n",
       "      <th>cds-NP_414546.1</th>\n",
       "      <th>cds-NP_414547.1</th>\n",
       "      <th>cds-NP_414548.1</th>\n",
       "      <th>cds-NP_414549.1</th>\n",
       "      <th>cds-NP_414550.1</th>\n",
       "      <th>cds-NP_414551.1</th>\n",
       "      <th>cds-NP_414552.1</th>\n",
       "      <th>...</th>\n",
       "      <th>cds-NP_418811.2</th>\n",
       "      <th>cds-NP_418812.1</th>\n",
       "      <th>cds-NP_418813.1</th>\n",
       "      <th>cds-NP_418814.1</th>\n",
       "      <th>cds-NP_418815.1</th>\n",
       "      <th>cds-NP_418816.1</th>\n",
       "      <th>cds-NP_418817.1</th>\n",
       "      <th>cds-NP_418818.1</th>\n",
       "      <th>cds-NP_418819.1</th>\n",
       "      <th>cds-NP_418820.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NT12189</th>\n",
       "      <td>-1.396155</td>\n",
       "      <td>-1.086905</td>\n",
       "      <td>-1.036221</td>\n",
       "      <td>-1.127845</td>\n",
       "      <td>-0.170443</td>\n",
       "      <td>0.154180</td>\n",
       "      <td>0.743706</td>\n",
       "      <td>0.231421</td>\n",
       "      <td>-1.702605</td>\n",
       "      <td>2.666265</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.026376</td>\n",
       "      <td>0.058862</td>\n",
       "      <td>0.081481</td>\n",
       "      <td>0.324293</td>\n",
       "      <td>0.111350</td>\n",
       "      <td>0.357424</td>\n",
       "      <td>3.099915</td>\n",
       "      <td>0.239958</td>\n",
       "      <td>0.529843</td>\n",
       "      <td>0.288563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12177</th>\n",
       "      <td>-2.021794</td>\n",
       "      <td>-1.487570</td>\n",
       "      <td>-1.481607</td>\n",
       "      <td>-0.445211</td>\n",
       "      <td>0.236060</td>\n",
       "      <td>-0.659636</td>\n",
       "      <td>-0.199121</td>\n",
       "      <td>0.661353</td>\n",
       "      <td>-0.481898</td>\n",
       "      <td>0.949621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.682059</td>\n",
       "      <td>-0.349514</td>\n",
       "      <td>-0.762992</td>\n",
       "      <td>-0.336101</td>\n",
       "      <td>-0.088115</td>\n",
       "      <td>0.407631</td>\n",
       "      <td>0.086031</td>\n",
       "      <td>0.108566</td>\n",
       "      <td>0.537271</td>\n",
       "      <td>0.573327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12130</th>\n",
       "      <td>-0.956499</td>\n",
       "      <td>-0.586719</td>\n",
       "      <td>-0.499014</td>\n",
       "      <td>-1.345601</td>\n",
       "      <td>0.244311</td>\n",
       "      <td>-0.234788</td>\n",
       "      <td>0.440232</td>\n",
       "      <td>0.620264</td>\n",
       "      <td>-1.095199</td>\n",
       "      <td>0.796310</td>\n",
       "      <td>...</td>\n",
       "      <td>0.574220</td>\n",
       "      <td>0.061330</td>\n",
       "      <td>-0.116296</td>\n",
       "      <td>0.051911</td>\n",
       "      <td>0.330066</td>\n",
       "      <td>0.430785</td>\n",
       "      <td>-0.603411</td>\n",
       "      <td>0.244223</td>\n",
       "      <td>-0.138190</td>\n",
       "      <td>0.107360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 4201 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene-id  cds-NP_414543.1  cds-NP_414544.1  cds-NP_414545.1  cds-NP_414546.1  \\\n",
       "NT12189        -1.396155        -1.086905        -1.036221        -1.127845   \n",
       "NT12177        -2.021794        -1.487570        -1.481607        -0.445211   \n",
       "NT12130        -0.956499        -0.586719        -0.499014        -1.345601   \n",
       "\n",
       "gene-id  cds-NP_414547.1  cds-NP_414548.1  cds-NP_414549.1  cds-NP_414550.1  \\\n",
       "NT12189        -0.170443         0.154180         0.743706         0.231421   \n",
       "NT12177         0.236060        -0.659636        -0.199121         0.661353   \n",
       "NT12130         0.244311        -0.234788         0.440232         0.620264   \n",
       "\n",
       "gene-id  cds-NP_414551.1  cds-NP_414552.1  ...  cds-NP_418811.2  \\\n",
       "NT12189        -1.702605         2.666265  ...        -0.026376   \n",
       "NT12177        -0.481898         0.949621  ...        -0.682059   \n",
       "NT12130        -1.095199         0.796310  ...         0.574220   \n",
       "\n",
       "gene-id  cds-NP_418812.1  cds-NP_418813.1  cds-NP_418814.1  cds-NP_418815.1  \\\n",
       "NT12189         0.058862         0.081481         0.324293         0.111350   \n",
       "NT12177        -0.349514        -0.762992        -0.336101        -0.088115   \n",
       "NT12130         0.061330        -0.116296         0.051911         0.330066   \n",
       "\n",
       "gene-id  cds-NP_418816.1  cds-NP_418817.1  cds-NP_418818.1  cds-NP_418819.1  \\\n",
       "NT12189         0.357424         3.099915         0.239958         0.529843   \n",
       "NT12177         0.407631         0.086031         0.108566         0.537271   \n",
       "NT12130         0.430785        -0.603411         0.244223        -0.138190   \n",
       "\n",
       "gene-id  cds-NP_418820.1  \n",
       "NT12189         0.288563  \n",
       "NT12177         0.573327  \n",
       "NT12130         0.107360  \n",
       "\n",
       "[3 rows x 4201 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deg = pd.read_csv('/Users/bamudamaris/Library/CloudStorage/GoogleDrive-fufordamaris@gmail.com/My Drive/Data/Datasets/deseqoutput_protein-ids.tsv', sep = '\\t', index_col = 0).T\n",
    "deg.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd186a8d-2060-412b-a8e8-cd7e2d544b4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete. Check the 'corr-plots' directory for the generated plots.\n"
     ]
    }
   ],
   "source": [
    "import warnings \n",
    "\n",
    "pearson = {}\n",
    "spearman = {}\n",
    "regress = {}\n",
    "\n",
    "def safe_correlation(x, y):\n",
    "    if len(set(x)) == 1 or len(set(y)) == 1:\n",
    "        return 0, 1  # Return correlation of 0 with p-value of 1 for constant input\n",
    "    return sp.stats.pearsonr(x, y)\n",
    "\n",
    "def safe_regression(x, y):\n",
    "    if len(set(x)) == 1 or len(set(y)) == 1:\n",
    "        return [0, 0, 0, 1, 0]  # Return dummy values for constant input\n",
    "    return stats.linregress(x, y)\n",
    "\n",
    "for var in df_new.columns:\n",
    "    g_var = 'cds-' + var.split('-')[1]\n",
    "    var_df = f_df[f_df['gene_name'] == var]\n",
    "    samples = set(var_df['k-samples'].str.cat(sep=',').split(','))\n",
    "    \n",
    "    if g_var in deg.columns:\n",
    "        # Merge predicted transcription rate and actual expression\n",
    "        d = pd.merge(df_new[var], deg[g_var], left_index=True, right_index=True, how='inner').dropna()\n",
    "        \n",
    "        # Filter to include only strains with the variant\n",
    "        d_variant = d[d.index.isin(samples)]\n",
    "        \n",
    "        if d_variant.empty:\n",
    "            print(f\"No data points with variant for {var}. Skipping.\")\n",
    "            continue\n",
    "        \n",
    "        d_variant.columns = ['Predicted transcription rate', 'log2foldchange']\n",
    "        \n",
    "        # Calculate correlations and regression\n",
    "        with warnings.catch_warnings():\n",
    "            warnings.simplefilter(\"ignore\")\n",
    "            p, pp = safe_correlation(d_variant['Predicted transcription rate'], d_variant['log2foldchange'])\n",
    "            s, ps = safe_correlation(sp.stats.rankdata(d_variant['Predicted transcription rate']), \n",
    "                                     sp.stats.rankdata(d_variant['log2foldchange']))\n",
    "            slope, intercept, r_value, p_value, std_err = safe_regression(d_variant['Predicted transcription rate'], \n",
    "                                                                          d_variant['log2foldchange'])\n",
    "        \n",
    "        pearson[var] = round(p, 2), pp\n",
    "        spearman[var] = round(s, 2), ps\n",
    "        regress[var] = [round(r_value, 2), p_value]\n",
    "        \n",
    "        # Create scatter plot\n",
    "        plt.figure(figsize=(5, 4))\n",
    "        plt.rcParams.update({'font.size': 14})\n",
    "        sns.scatterplot(data=d_variant, x='Predicted transcription rate', y='log2foldchange')\n",
    "        \n",
    "        if len(set(d_variant['Predicted transcription rate'])) > 1 and len(set(d_variant['log2foldchange'])) > 1:\n",
    "            sns.regplot(data=d_variant, x='Predicted transcription rate', y='log2foldchange', \n",
    "                        scatter=False, \n",
    "                        label=f'r={p:.2f}')\n",
    "        else:\n",
    "            plt.plot([], [], color='red', label='Constant input')\n",
    "        \n",
    "        plt.ylabel( '$\\log_2$(FC)\\n(Actual Expression)')\n",
    "        plt.xlabel('Predicted transcription rate\\nrelative to $\\it{E. coli}$ K-12')\n",
    "        plt.title(f\"{g_var}\")\n",
    "        plt.tight_layout()\n",
    "        plt.legend()\n",
    "        plt.savefig(f'corr-plots_var_present/{var}_variant_only.svg', dpi=300)\n",
    "        plt.close()\n",
    "\n",
    "print(\"Analysis complete. Check the 'corr-plots' directory for the generated plots.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb6f3560-1c0f-4461-9c52-d52bfef1e1a5",
   "metadata": {},
   "source": [
    "Comparing predicted expression rates of strains with the variants vs those without the variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "162d2442-b352-4c38-9cdc-b53a6616a981",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete. Check the 'corr-plots' directory for the generated plots.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "pearson = {}\n",
    "spearman = {}\n",
    "regress = {}\n",
    "\n",
    "for var in df_new.columns:\n",
    "    g_var = 'cds-' + var.split('-')[1]\n",
    "    var_df = f_df[f_df['gene_name'] == var]\n",
    "    samples = set(var_df['k-samples'].str.cat(sep=',').split(','))\n",
    "    \n",
    "    if g_var in deg.columns:\n",
    "        # Merge predicted transcription rate and actual expression\n",
    "        d = pd.merge(df_new[var], deg[g_var], left_index=True, right_index=True, how='inner').dropna()\n",
    "        \n",
    "        # Add a column to indicate variant presence\n",
    "        d['variant_present'] = d.index.isin(samples).astype(int)\n",
    "        d.columns = ['Predicted transcription rate', 'log2foldchange', 'variant_present']\n",
    "        \n",
    "        # Separate dataframes for variant present and absent\n",
    "        d_variant = d[d['variant_present'] == 1]\n",
    "        d_no_variant = d[d['variant_present'] == 0]\n",
    "        \n",
    "        # Calculate correlations and regression for both cases\n",
    "        for case, data in [(\"Variant\", d_variant), (\"No Variant\", d_no_variant)]:\n",
    "            if not data.empty:\n",
    "                p, pp = sp.stats.pearsonr(data['Predicted transcription rate'], data['log2foldchange'])\n",
    "                s, ps = sp.stats.spearmanr(data['Predicted transcription rate'], data['log2foldchange'])\n",
    "                slope, intercept, r_value, p_value, std_err = stats.linregress(data['Predicted transcription rate'], data['log2foldchange'])\n",
    "                \n",
    "                pearson[f\"{var}_{case}\"] = round(p, 2), pp\n",
    "                spearman[f\"{var}_{case}\"] = round(s, 2), ps\n",
    "                regress[f\"{var}_{case}\"] = [round(r_value, 2), p_value]\n",
    "        \n",
    "        # Create scatter plot\n",
    "        plt.figure(figsize=(10, 6))\n",
    "        sns.scatterplot(data=d, x='Predicted transcription rate', y='log2foldchange', \n",
    "                        hue='variant_present', palette={0: 'blue', 1: 'red'}, alpha=0.7)\n",
    "        \n",
    "        # Add regression lines for both cases\n",
    "        for case, data, color in [(\"Variant\", d_variant, 'red'), (\"No Variant\", d_no_variant, 'blue')]:\n",
    "            if not data.empty:\n",
    "                sns.regplot(data=data, x='Predicted transcription rate', y='log2foldchange', \n",
    "                            scatter=False, color=color, label=f'{case} (r={pearson[f\"{var}_{case}\"][0]:.2f}, p={pearson[f\"{var}_{case}\"][1]:.3f})')\n",
    "        \n",
    "        plt.ylabel('log2foldchange (Actual Expression)')\n",
    "        plt.xlabel('Predicted transcription rate')\n",
    "        plt.title(f\"{g_var}\")\n",
    "        plt.legend(title='Variant present')\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f'corr-plots/{var}_comparison.svg', dpi=300)\n",
    "        plt.close()\n",
    "\n",
    "print(\"Analysis complete. Check the 'corr-plots' directory for the generated plots.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3d039aa3-a162-444e-8323-c63893305ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save computed correlation values to dataframe for downstream analysis\n",
    "df_p = pd.DataFrame(pearson).T.rename(columns = {0: 'pearson', 1: 'pearson(p-value)'})\n",
    "df_sp = pd.DataFrame(spearman).T.rename(columns = {0:'spearman', 1:'spearman(p-value)'})\n",
    "linregress = pd.DataFrame(regress).T.rename(columns = {0: 'r_value', 1: 'regress(p-value)'})    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0506d6c2-ea59-4611-84ca-3d44092a954d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>association</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_418030.1-793</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417082.1-613</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415117.2-501</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416726.1-2615</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417082.1-607</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415592.1-4200</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415129.1-35</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415592.1-4214</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416770.1-9</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416834.1-4027</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4904 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      association\n",
       "cds-NP_418030.1-793             0\n",
       "cds-NP_417082.1-613             0\n",
       "cds-NP_415117.2-501             0\n",
       "cds-NP_416726.1-2615            0\n",
       "cds-NP_417082.1-607             0\n",
       "...                           ...\n",
       "cds-NP_415592.1-4200            0\n",
       "cds-NP_415129.1-35              0\n",
       "cds-NP_415592.1-4214            0\n",
       "cds-NP_416770.1-9               0\n",
       "cds-NP_416834.1-4027            0\n",
       "\n",
       "[4904 rows x 1 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outcomes = {}\n",
    "for i in tx_rates.columns:\n",
    "    name = i + '.tsv'\n",
    "    a = pd.read_csv('/Users/bamudamaris/ownCloud2/model/ecoli/validation_2024/tested_patterns/' + name,  sep = '\\t', index_col = 0, names = ['pattern', 'lrt-pvalue', 'beta', 'k-samples'])\n",
    "    a['assigned'] = a['lrt-pvalue'].apply(lambda x: 1 if x <= 6.95E-06 else 0)\n",
    "    outcomes[i] = a['assigned'].values.tolist()\n",
    "outcome = pd.DataFrame(outcomes).T.rename(columns = {0: 'association'}) \n",
    "outcome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "eb0e07a8-3ed0-475f-b57b-bb2a10ff92f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.151193e-01</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9.478283e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.448540e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.237597e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9.302139e-01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.966887e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>5.351611e-01</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.387368e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>0.21</td>\n",
       "      <td>5.638126e-02</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>3.137891e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418085.2-2294</th>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>0.31</td>\n",
       "      <td>5.926000e-02</td>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.465319e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417610.1-1604</th>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4.008327e-21</td>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.889370e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>2.683375e-02</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>4.712333e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.280097e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.488176e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>6.809198e-03</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.856504e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pearson  pearson(p-value)  spearman  spearman(p-value)  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01      0.02       8.151193e-01   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01     -0.04       6.448540e-01   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01      0.01       9.302139e-01   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01     -0.06       5.351611e-01   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01      0.21       5.638126e-02   \n",
       "...                       ...               ...       ...                ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02      0.31       5.926000e-02   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18      0.74       4.008327e-21   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05     -0.21       2.683375e-02   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01     -0.04       7.280097e-01   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02     -0.39       6.809198e-03   \n",
       "\n",
       "                      r_value  regress(p-value)  association  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01            1   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01            1   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01            1   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01            1   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01            1   \n",
       "...                       ...               ...          ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02            1   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18            1   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05            1   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01            1   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02            1   \n",
       "\n",
       "                      adjusted_p_values  \n",
       "cds-NP_414578.2-75         9.478283e-01  \n",
       "cds-NP_418204.2-940        8.237597e-01  \n",
       "cds-NP_414880.2-2266       6.966887e-01  \n",
       "cds-NP_418208.1-940        6.387368e-01  \n",
       "cds-NP_416773.1-1254       3.137891e-01  \n",
       "...                                 ...  \n",
       "cds-NP_418085.2-2294       1.465319e-01  \n",
       "cds-NP_417610.1-1604       1.889370e-16  \n",
       "cds-NP_418554.1-1855       4.712333e-04  \n",
       "cds-NP_418686.1-718        8.488176e-01  \n",
       "cds-YP_588463.1-2320       1.856504e-01  \n",
       "\n",
       "[267 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#combine dataframes\n",
    "# prediction = pd.concat([df_p, df_sp, linregress, outcome], axis = 1).dropna()\n",
    "# prediction['association'] = prediction['association'].astype(int)\n",
    "prediction = pd.read_csv('/Users/bamudamaris/ownCloud2/model/ecoli/tx-panfeed_230_rerun/asso-correlation-panfeed230-sig.tsv', sep = '\\t', index_col = 0)\n",
    "prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cab872bd-87b0-4647-a51f-13eb78d5e948",
   "metadata": {},
   "source": [
    "Focus on significantly associated variants only "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ef92125b-9fa4-40ee-813f-323ae90bff81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.151193e-01</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9.478283e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.448540e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.237597e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9.302139e-01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.966887e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>5.351611e-01</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.387368e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>0.21</td>\n",
       "      <td>5.638126e-02</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>3.137891e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418085.2-2294</th>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>0.31</td>\n",
       "      <td>5.926000e-02</td>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.465319e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417610.1-1604</th>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4.008327e-21</td>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.889370e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>2.683375e-02</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>4.712333e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.280097e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.488176e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>6.809198e-03</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.856504e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pearson  pearson(p-value)  spearman  spearman(p-value)  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01      0.02       8.151193e-01   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01     -0.04       6.448540e-01   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01      0.01       9.302139e-01   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01     -0.06       5.351611e-01   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01      0.21       5.638126e-02   \n",
       "...                       ...               ...       ...                ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02      0.31       5.926000e-02   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18      0.74       4.008327e-21   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05     -0.21       2.683375e-02   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01     -0.04       7.280097e-01   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02     -0.39       6.809198e-03   \n",
       "\n",
       "                      r_value  regress(p-value)  association  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01            1   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01            1   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01            1   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01            1   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01            1   \n",
       "...                       ...               ...          ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02            1   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18            1   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05            1   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01            1   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02            1   \n",
       "\n",
       "                      adjusted_p_values  \n",
       "cds-NP_414578.2-75         9.478283e-01  \n",
       "cds-NP_418204.2-940        8.237597e-01  \n",
       "cds-NP_414880.2-2266       6.966887e-01  \n",
       "cds-NP_418208.1-940        6.387368e-01  \n",
       "cds-NP_416773.1-1254       3.137891e-01  \n",
       "...                                 ...  \n",
       "cds-NP_418085.2-2294       1.465319e-01  \n",
       "cds-NP_417610.1-1604       1.889370e-16  \n",
       "cds-NP_418554.1-1855       4.712333e-04  \n",
       "cds-NP_418686.1-718        8.488176e-01  \n",
       "cds-YP_588463.1-2320       1.856504e-01  \n",
       "\n",
       "[267 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#perform multiple test corrections\n",
    "\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "sig = prediction[prediction['association'] == 1]\n",
    "\n",
    "\n",
    "# Perform multiple test correlation using Benjamini-Hochberg (FDR) correction\n",
    "sig['adjusted_p_values'] = multipletests(sig['pearson(p-value)'], method='fdr_bh')[1]\n",
    "\n",
    "#sig.to_csv('asso-correlation-panfeed230-sig.tsv', sep = '\\t')\n",
    "\n",
    "sig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5b0bb355-1738-4aa0-b627-8867f3afa0c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For threshold 0.001, number of significant correlations is: 22 (8.24%)\n",
      "For threshold 0.002, number of significant correlations is: 22 (8.24%)\n",
      "For threshold 0.003, number of significant correlations is: 25 (9.36%)\n",
      "For threshold 0.004, number of significant correlations is: 25 (9.36%)\n",
      "For threshold 0.005, number of significant correlations is: 25 (9.36%)\n",
      "For threshold 0.0001, number of significant correlations is: 21 (7.87%)\n",
      "For threshold 0.01, number of significant correlations is: 26 (9.74%)\n",
      "For threshold 0.05, number of significant correlations is: 34 (12.73%)\n"
     ]
    }
   ],
   "source": [
    "thresholds = [0.001,0.002, 0.003, 0.004, 0.005, 0.0001, 0.01, 0.05]  \n",
    "\n",
    "total_var = len(sig)\n",
    "\n",
    "for threshold in thresholds:\n",
    "    significant_correlations = sig[(sig['pearson'] > 0) & (sig['adjusted_p_values'] <= threshold)]\n",
    "    num_significant = len(significant_correlations)\n",
    "    percentage = (num_significant / total_var) * 100\n",
    "    print(f\"For threshold {threshold}, number of significant correlations is: {num_significant} ({percentage:.2f}%)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "04b8d1ca-8f9e-412b-bc68-070046e53dbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEGCAYAAABiq/5QAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAYA0lEQVR4nO3de5RdZX3G8e/jkAQUNFymOE0ICYhSltWgI17wAhEUlaVQUaFqQ4uNN7zhJVhcFV11FWoVrbZqBCW2louRCEbFIgmKVsCJhpAQhYAgSXMZFQpqjcn46x/7HbI5OTNzZubsfWbmfT5rnTV7v/v2m53JM3v25d2KCMzMLB+P6nQBZmZWLwe/mVlmHPxmZplx8JuZZcbBb2aWmb06XUArDjrooJg7d26nyzAzm1RWr179y4jobmyfFME/d+5c+vr6Ol2GmdmkIuneZu0+1WNmlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv00JPbPnIGlMn57Zczqy3fFu22ysJkWXDWYj2br5Pg5dvGJMy9574ckd2e54t202Vj7iNzPLjIPfzCwzDn4zs8w4+M3MMlN58EvqkvQTSSvS+DxJN0vaKOkKSdOrrsHMzHar44j/HcCG0viFwEUR8QTgfuCsGmowM7Ok0uCXNBt4GXBxGhewAFiWZlkKnFJlDWZm9khVH/F/Angf8Mc0fiDwQETsSuObgFkV12BmZiWVBb+kk4HtEbF6jMsvktQnqa+/v7/N1ZmZ5avKI/5jgZdLuge4nOIUzyeBmZIGnxieDWxutnBELImI3ojo7e7e4yXxZmY2RpUFf0S8PyJmR8Rc4HRgZUS8FlgFnJZmWwhcXVUNZma2p07cx78YOEfSRopz/pd0oAYzs2zV0klbRNwA3JCG7waOqWO7Zma2Jz+5axPGeLo4NrPWuVtmmzA61bWyWW58xG9mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv5lZZqp82frekm6RdKuk9ZI+lNovlfRzSWvSZ35VNZiZ2Z6q7I9/B7AgIn4jaRrwfUnfStPeGxHLKty2mZkNobLgj4gAfpNGp6VPVLU9MzNrTaXn+CV1SVoDbAeui4ib06SPSFor6SJJM4ZYdpGkPkl9/f39VZZpZpaVSoM/IgYiYj4wGzhG0pOB9wNHAs8ADgAWD7HskojojYje7u7uKss0M8tKLXf1RMQDwCrgpIjYEoUdwBeBY+qowczMClXe1dMtaWYa3gc4EfippJ7UJuAUYF1VNZiZ2Z6qvKunB1gqqYviF8yVEbFC0kpJ3YCANcCbKqzBzMwaVHlXz1rg6CbtC6raptmYdE2j+APULA9VHvGbTQ4DOzl08YoxLXrvhSe3uRiz6rnLBjOzzDj4zcwy4+A3M8uMg9/MLDMOfjOzzDj4zcwy4+A3M8uMg9/MLDMOfjOzzDj4zcwy4+A3M8uMg9/MLDMOfjOzzDj4zcwy4+A3M8tMla9e3FvSLZJulbRe0odS+zxJN0vaKOkKSdOrqsHMzPZU5RH/DmBBRDwVmA+cJOlZwIXARRHxBOB+4KwKazAzswaVBX8UfpNGp6VPAAuAZal9KcUL183MrCaVnuOX1CVpDbAduA64C3ggInalWTYBs4ZYdpGkPkl9/f39VZZpZpaVSoM/IgYiYj4wGzgGOHIUyy6JiN6I6O3u7q6qRDOz7NRyV09EPACsAp4NzJQ0+JL32cDmOmowM7NClXf1dEuamYb3AU4ENlD8AjgtzbYQuLqqGszMbE97jTzLmPUASyV1UfyCuTIiVki6Hbhc0j8APwEuqbAGMzNrUFnwR8Ra4Ogm7XdTnO83M7MO8JO7ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlpkq38B1iKRVkm6XtF7SO1L7+ZI2S1qTPi+tqgYzM9tTlW/g2gW8OyJ+LGk/YLWk69K0iyLinyvctpmZDaHKN3BtAbak4YckbQBmVbU9MzNrTS3n+CXNpXgN482p6WxJayV9QdL+ddRgZmaFyoNf0r7AV4F3RsSDwGeAw4H5FH8RfGyI5RZJ6pPU19/fX3WZZmbZqDT4JU2jCP0vR8RVABGxLSIGIuKPwOcZ4sXrEbEkInojore7u7vKMs3MstJS8Es6tpW2hukCLgE2RMTHS+09pdlOBda1VqqZmbVDqxd3PwU8rYW2smOB1wO3SVqT2v4OOEPSfCCAe4A3tliDmZm1wbDBL+nZwHOAbknnlCY9FugabtmI+D6gJpO+OdoizcysfUY64p8O7Jvm26/U/iBwWlVFmZlZdYYN/oj4LvBdSZdGxL011WRmZhVq9Rz/DElLgLnlZSJiQRVFmZlZdVoN/q8AnwUuBgaqK8fMzKrWavDviojPVFqJmZnVotUHuL4u6S2SeiQdMPiptDIzM6tEq0f8C9PX95baAjisveWYmVnVWgr+iJhXdSFmZlaPloJf0l81a4+IL7W3HDMzq1qrp3qeURreG3gh8GPAwW9mNsm0eqrnbeVxSTOBy6soyMzMqjXWbpl/C/i8v5nZJNTqOf6vU9zFA0XnbH8GXFlVUWZmVp1Wz/GXX4y+C7g3IjZVUI+ZmVWspVM9qbO2n1L00Lk/8IcqizLLRtc0JI3p0zN7Tqert0mq1VM9rwY+CtxA0cf+pyS9NyKWVVib2dQ3sJNDF68Y06L3Xnhym4uxXLR6quc84BkRsR1AUjfwHWDI4Jd0CMXtngdTXB9YEhGfTF09XEHR0+c9wKsj4v6xfgNmZjY6rd7V86jB0E9+1cKyu4B3R8RRwLOAt0o6CjgXuD4ijgCuT+NmZlaTVo/4r5X0beCyNP4aRniFYkRsAbak4YckbQBmAa8AjkuzLaU4fbR4VFWbmdmYjfTO3ScAB0fEeyX9BfDcNOmHwJdb3YikucDRwM1pfVvSpK0Up4KaLbMIWAQwZ44vYpmZtctIp2s+QfF+XSLiqog4JyLOAZanaSOStC/wVeCdEfFgeVpEBLufD6Bh2pKI6I2I3u7u7lY2ZWZmLRgp+A+OiNsaG1Pb3JFWLmkaReh/OSKuSs3bJPWk6T3A9qGWNzOz9hsp+GcOM22f4RaUJOASYENEfLw06Rp29++/ELh6hBrMzKyNRgr+Pkl/29go6Q3A6hGWPRZ4PbBA0pr0eSlwAXCipDuBE9K4mZnVZKS7et4JLJf0WnYHfS8wHTh1uAUj4vsUD3s188JR1GhmZm00bPBHxDbgOZKOB56cmr8RESsrr8zMzCrRan/8q4BVFddiZmY1GGt//GZmNkk5+M3MMuPgNzPLjIPfzCwzDn4zs8w4+M3MMuPgNzPLjIPfzCwzDn4zs8w4+M3MMuPgNzPLjIPfzCwzDn5rm57Zc5A05o+Z1aOl3jnNWrF1830cunjFmJe/98KT21iNmQ2lsiN+SV+QtF3SulLb+ZI2N7yRy8zMalTlqZ5LgZOatF8UEfPT55sVbt/MzJqoLPgj4nvAr6tav5mZjU0nLu6eLWltOhW0/1AzSVokqU9SX39/f531mZlNaXUH/2eAw4H5wBbgY0PNGBFLIqI3Inq7u7trKs/MbOqrNfgjYltEDETEH4HPA8fUuX0zM6s5+CX1lEZPBdYNNa+ZmVWjsvv4JV0GHAccJGkT8EHgOEnzgQDuAd5Y1fbNzKy5yoI/Is5o0nxJVdszM7PWuMsGM7PMOPjNzDLj4Dczy4yD38wsMw5+M7PMOPjNzDLj4Dczy4yD38wsMw5+M7PMOPjNzDLj4Dczy4yD38wsMw5+M7PMOPjNzDLj4Dczy0xlwZ9epr5d0rpS2wGSrpN0Z/o65MvWzcysGlUe8V8KnNTQdi5wfUQcAVyfxs3MrEaVBX9EfA/4dUPzK4ClaXgpcEpV2zczs+bqPsd/cERsScNbgYOHmlHSIkl9kvr6+/vrqc7omT0HSWP6mNnkUNk7d0cSESEphpm+BFgC0NvbO+R81l5bN9/HoYtXjGnZey88uc3VmFkV6j7i3yapByB93V7z9s3Msld38F8DLEzDC4Gra96+mVn2qryd8zLgh8CTJG2SdBZwAXCipDuBE9K4mZnVqLJz/BFxxhCTXljVNs2y0jVtzBfVu6bvzcAffj+mZR8/6xC2bPrFmJbtmT2HrZvvG9Oy49227daxi7tmNk4DO8d1Ib4TF/HHc/PAeLdtu7nLBjOzzDj4zcwy4+A3M8uMg9/MLDMOfjOzzDj4zcwy4+A3M8uMg9/MLDNTPvjH081wz+w5nS7fzKztpvyTu+5m2Mzskab8Eb+ZmT2Sg9/MLDMOfjOzzDj4pyC/N9ds4hjP/8eqbjKZ8hd3c+QL2mYTx0TsirojwS/pHuAhYADYFRG9najDzCxHnTziPz4iftnB7ZuZZcnn+M3MMtOp4A/gvyStlrSo2QySFknqk9TX399fc3lmZlNXp4L/uRHxNOAlwFslPb9xhohYEhG9EdHb3d1df4VmZlNUR4I/Ijanr9uB5cAxnajDzCxHtQe/pMdI2m9wGHgRsK7uOszMctWJu3oOBpanh4X2Av4zIq7tQB1mZlmqPfgj4m7gqXVvtxN6Zs9h6+b7xrTs42cdwpZNv2hzRWZt0DXNT3lPcn5yt0J+gtampIGd/rme5Hwfv5lZZhz8ZmaZcfCbmWXGwT9RpQto7lrZrD38/u3dfHF3ovIFNLO28s0Wu/mI38wsMw5+M7PMOPjNzDLj4Dczy4yD38wsMw5+M7PMOPjNzDLj4Dczy4wf4BqOu581m1g69X9yimWBg38443h6Fqbe035mHdepJ9qn2JP0HTnVI+kkST+TtFHSuZ2owcwsV514524X8K/AS4CjgDMkHVV3HWZmuerEEf8xwMaIuDsi/gBcDryiA3WYmWVJEVHvBqXTgJMi4g1p/PXAMyPi7Ib5FgGL0uiTgJ+Nc9MHAb8c5zo6wXXXy3XXb7LWPhnqPjQiuhsbJ+zF3YhYAixp1/ok9UVEb7vWVxfXXS/XXb/JWvtkrRs6c6pnM3BIaXx2ajMzsxp0Ivh/BBwhaZ6k6cDpwDUdqMPMLEu1n+qJiF2Szga+DXQBX4iI9TVsum2njWrmuuvluus3WWufrHXXf3HXzMw6y331mJllxsFvZpaZKRX8kg6QdJ2kO9PX/ZvMc7ykNaXP7yWdkqZdKunnpWnzJ0rdab6BUm3XlNrnSbo5dYFxRbpoPiHqljRf0g8lrZe0VtJrStNq3d8jdRUiaUbafxvT/pxbmvb+1P4zSS+uss4x1H2OpNvT/r1e0qGlaU1/ZiZI3WdK6i/V94bStIXp5+pOSQsnWN0XlWq+Q9IDpWkd29+jEhFT5gP8E3BuGj4XuHCE+Q8Afg08Oo1fCpw2UesGfjNE+5XA6Wn4s8CbJ0rdwBOBI9LwnwJbgJl172+KGwnuAg4DpgO3Akc1zPMW4LNp+HTgijR8VJp/BjAvradrAtV9fOln+M2DdQ/3MzNB6j4T+HSTZQ8A7k5f90/D+0+UuhvmfxvFDSod3d+j/UypI36Krh+WpuGlwCkjzH8a8K2I+F2VRbVgtHU/TEVfsQuAZWNZfpxGrDsi7oiIO9Pw/wDbgT2eJKxBK12FlL+fZcAL0/59BXB5ROyIiJ8DG9P6JkTdEbGq9DN8E8WzMZ02nq5ZXgxcFxG/joj7geuAkyqqs9Fo6z4DuKyWytpoqgX/wRGxJQ1vBQ4eYf7T2fMf7SPpT+aLJM1oe4XNtVr33pL6JN00eHoKOBB4ICJ2pfFNwKzqSn2EUe1vScdQHEXdVWqua3/PAu4rjTfbTw/Pk/bn/1Ls31aWrcpot30W8K3SeLOfmTq0Wvcr07//MkmDD3ZOiv2dTqnNA1aWmju1v0dlwnbZMBRJ3wEe32TSeeWRiAhJQ96rKqkH+HOK5wkGvZ8iwKZT3KO7GPjweGtO22tH3YdGxGZJhwErJd1GEU6VafP+/ndgYUT8MTVXtr9zJOl1QC/wglLzHj8zEXFX8zXU7uvAZRGxQ9IbKf7aWtDhmkbjdGBZRAyU2iby/n7YpAv+iDhhqGmStknqiYgtKWi2D7OqVwPLI2Jnad2DR687JH0ReE9biqY9dUfE5vT1bkk3AEcDXwVmStorHaW2tQuMdtQt6bHAN4DzIuKm0ror299NtNJVyOA8myTtBTwO+FWLy1alpW1LOoHil/ELImLHYPsQPzN1BNGIdUfEr0qjF1NcMxpc9riGZW9oe4XNjebf+nTgreWGDu7vUZlqp3quAQbvAFgIXD3MvHucm0vhNXje/BRgXftLbGrEuiXtP3gqRNJBwLHA7VFcUVpFcb1iyOUr0krd04HlwJciYlnDtDr3dytdhZS/n9OAlWn/XgOcnu76mQccAdxSYa2jqlvS0cDngJdHxPZSe9OfmQlUd09p9OXAhjT8beBFqf79gRfxyL/Mq9RSlzKSjqS48PzDUlsn9/fodPrqcjs/FOdjrwfuBL4DHJDae4GLS/PNpfgt/qiG5VcCt1EE0H8A+06UuoHnpNpuTV/PKi1/GEUQbQS+AsyYQHW/DtgJrCl95ndifwMvBe6gOAI7L7V9mCIwAfZO+29j2p+HlZY9Ly33M+AlNf9cj1T3d4Btpf17zUg/MxOk7n8E1qf6VgFHlpb9m/TvsBH464lUdxo/H7igYbmO7u/RfNxlg5lZZqbaqR4zMxuBg9/MLDMOfjOzzDj4zcwy4+A3M8uMg99qI+nxki6XdJek1ZK+KemJFWxnrqRhnwlI8/xlabxX0r+0u5ZWSLon3fdd1fp7JK0YYZ7pkr6XHlyzKc7Bb7VID2ktB26IiMMj4ukUXTaM1J/S4PJ7DTc+BnOBh4M/Ivoi4u3jXOdEdQ7w+eFmiKJDsuuB1ww3n00NDn6ry/HAzoj47GBDRNwaETeq8FFJ6yTdptRnv6TjJN2ool/z25uMd6XlfpQ6+npj40bTkf2Nkn6cPs9Jky4Anqei3/R3pXWvSMscIOlraZ03SXpKaj9f0hck3SDpbkl7/KKQ9CZJHy2Nnynp02n4a+kvnfWSFg1R67rS+HsknZ+GD5d0bVr+xvTkKJJelfbbrZK+N8S+fyVwbameq9P3cKekD5bm+xrw2iHWYVNJp58g8yePD/B24KIhpr2SouvdLoq/AH4B9FD01/JbYF6ar3F8EfCBNDwD6KPoLXEusC61PxrYOw0fAfSV1rWiVMPD48CngA+m4QXAmjR8PvDfaVsHUfTjM63he+mm6NZ3cPxbwHPT8OCTzftQPK18YBq/J63v4bpT+3uA89Pw9ex+r8EzKbqTgOIJ0VlpeGaTfTsPWF0aP5PinQgHluroTdO6gP5O/6z4U/3H5/NsInguRS+NA8A2Sd8FngE8CNwSRR/4g8rjLwKeImmwn6LHUYT7HaX5pwGfVvF2rwGKF8O0Us8rASJipaQDVXQ0B/CNKDpB2yFpO8Uvqk2DC0ZEf/pr4FkUXVkcCfwgTX67pFPT8CGp1nJHZU1J2peiO4CvFGfMgOKXD2ndl0q6EriqyeI9QH9D23WROkiTdFX6fvsiYkDSHyTtFxEPjVSXTV4OfqvLenZ3JDcavx1mXMDbIuIRHXip9MpE4F0U/dg8leLU5u/HUEPZjtLwAM3/D11O0fvrTyl6gA1JxwEnAM+OiN+p6Llx74bldvHI06+D0x9F8c6F+Y0biog3SXom8DJgtaSnxyN7vfy/Jttp7KelPD6D8e8jm+B8jt/qshKYUT63Lekpkp4H3Ai8Jp2z7waeT2u9X34beLOkaWl9T5T0mIZ5HgdsieIdAK+nOJ0B8BCw3xDrvZF0rjsF9i8j4sEW6hm0nOKtTWdQ/BIYrOP+FPpHAs9qstw24E/SXxgzgJMB0rZ/LulVqSZJemoaPjwibo6Iv6c4sj+kYZ13UJxCKjsxXcfYh6JX1B+kdR2Yvted2JTm4LdaREQApwInqLidcz1F74xbKYJyLUWvhiuB90XE1hZWezFFt7c/ThdFP8eeR+D/BiyUdCvFaZfBvxjWAgPpoui7GpY5H3i6pLUUF4EXMgpRvC5wA8VLOQZ/gV0L7CVpQ1rnTU2W20nRC+QtFNc8flqa/FrgrPR9rGf36wA/mi6Ir6O4/nBrwzp/C9wl6Qml5lso3uOwFvhqRPSl9uMp3ptgU5x75zSb4tJ1hadHxAcknUlxMffsJvNdBZwbEXc0TrOpxef4zaa4iFieTuMMScVLR77m0M+Dj/jNzDLjc/xmZplx8JuZZcbBb2aWGQe/mVlmHPxmZpn5f+cgkXzkAj3aAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sig_filtered = sig[sig['adjusted_p_values'] < 0.05]\n",
    "plt.hist(sig['pearson'], bins=20, edgecolor='black')\n",
    "plt.xlabel('Correlation values (p)')\n",
    "plt.ylabel('Count')\n",
    "plt.rcParams.update({'font.size': 12})\n",
    "#plt.title('Associated variants')\n",
    "#plt.savefig('sig_correlation.svg', dpi = 300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5af44804-0ca5-4224-9073-c87b97d739e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.151193e-01</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9.478283e-01</td>\n",
       "      <td>cds-NP_414578.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.448540e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.237597e-01</td>\n",
       "      <td>cds-NP_418204.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9.302139e-01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.966887e-01</td>\n",
       "      <td>cds-NP_414880.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>5.351611e-01</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.387368e-01</td>\n",
       "      <td>cds-NP_418208.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>0.21</td>\n",
       "      <td>5.638126e-02</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>3.137891e-01</td>\n",
       "      <td>cds-NP_416773.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418085.2-2294</th>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>0.31</td>\n",
       "      <td>5.926000e-02</td>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.465319e-01</td>\n",
       "      <td>cds-NP_418085.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417610.1-1604</th>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4.008327e-21</td>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.889370e-16</td>\n",
       "      <td>cds-NP_417610.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>2.683375e-02</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>4.712333e-04</td>\n",
       "      <td>cds-NP_418554.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.280097e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.488176e-01</td>\n",
       "      <td>cds-NP_418686.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>6.809198e-03</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.856504e-01</td>\n",
       "      <td>cds-YP_588463.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pearson  pearson(p-value)  spearman  spearman(p-value)  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01      0.02       8.151193e-01   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01     -0.04       6.448540e-01   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01      0.01       9.302139e-01   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01     -0.06       5.351611e-01   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01      0.21       5.638126e-02   \n",
       "...                       ...               ...       ...                ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02      0.31       5.926000e-02   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18      0.74       4.008327e-21   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05     -0.21       2.683375e-02   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01     -0.04       7.280097e-01   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02     -0.39       6.809198e-03   \n",
       "\n",
       "                      r_value  regress(p-value)  association  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01            1   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01            1   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01            1   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01            1   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01            1   \n",
       "...                       ...               ...          ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02            1   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18            1   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05            1   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01            1   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02            1   \n",
       "\n",
       "                      adjusted_p_values             gene  \n",
       "cds-NP_414578.2-75         9.478283e-01  cds-NP_414578.2  \n",
       "cds-NP_418204.2-940        8.237597e-01  cds-NP_418204.2  \n",
       "cds-NP_414880.2-2266       6.966887e-01  cds-NP_414880.2  \n",
       "cds-NP_418208.1-940        6.387368e-01  cds-NP_418208.1  \n",
       "cds-NP_416773.1-1254       3.137891e-01  cds-NP_416773.1  \n",
       "...                                 ...              ...  \n",
       "cds-NP_418085.2-2294       1.465319e-01  cds-NP_418085.2  \n",
       "cds-NP_417610.1-1604       1.889370e-16  cds-NP_417610.1  \n",
       "cds-NP_418554.1-1855       4.712333e-04  cds-NP_418554.1  \n",
       "cds-NP_418686.1-718        8.488176e-01  cds-NP_418686.1  \n",
       "cds-YP_588463.1-2320       1.856504e-01  cds-YP_588463.1  \n",
       "\n",
       "[267 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig['gene'] = ['cds-' + g.split('-')[1] for g in sig.index]\n",
    "sig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "911c3813-9d9f-4541-972c-ecb174d001f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.898129</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.815119</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.898129</td>\n",
       "      <td>1</td>\n",
       "      <td>0.947828</td>\n",
       "      <td>cds-NP_414578.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.697265</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.644854</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.697265</td>\n",
       "      <td>1</td>\n",
       "      <td>0.823760</td>\n",
       "      <td>cds-NP_418204.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <td>0.07</td>\n",
       "      <td>0.527083</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.930214</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.527083</td>\n",
       "      <td>1</td>\n",
       "      <td>0.696689</td>\n",
       "      <td>cds-NP_414880.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pearson  pearson(p-value)  spearman  spearman(p-value)  \\\n",
       "cds-NP_414578.2-75      -0.01          0.898129      0.02           0.815119   \n",
       "cds-NP_418204.2-940     -0.04          0.697265     -0.04           0.644854   \n",
       "cds-NP_414880.2-2266     0.07          0.527083      0.01           0.930214   \n",
       "\n",
       "                      r_value  regress(p-value)  association  \\\n",
       "cds-NP_414578.2-75      -0.01          0.898129            1   \n",
       "cds-NP_418204.2-940     -0.04          0.697265            1   \n",
       "cds-NP_414880.2-2266     0.07          0.527083            1   \n",
       "\n",
       "                      adjusted_p_values             gene  \n",
       "cds-NP_414578.2-75             0.947828  cds-NP_414578.2  \n",
       "cds-NP_418204.2-940            0.823760  cds-NP_418204.2  \n",
       "cds-NP_414880.2-2266           0.696689  cds-NP_414880.2  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = sig.copy()\n",
    "data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5d8fec99-3036-40a0-aad8-282f442bcf49",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = sig.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d0fa824-ef79-42a2-a944-151a43a2b49d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No correlation (r = 0.0): 5 (1.87%)\n",
      "Weak positive (0 < r ≤ 0.2): 83 (31.09%)\n",
      "Weak negative (-0.2 ≤ r < 0): 92 (34.46%)\n",
      "Moderate positive (0.2 < r < 0.8): 39 (14.61%)\n",
      "Moderate negative (-0.8 < r < -0.2): 41 (15.36%)\n",
      "Strong positive (r ≥ 0.8): 7 (2.62%)\n",
      "Strong negative (r ≤ -0.8): 0 (0.00%)\n",
      "\n",
      "Correlations > 0.2: 46 (17.23%)\n",
      "Correlations < -0.2: 41 (15.36%)\n"
     ]
    }
   ],
   "source": [
    "# Total number of correlations\n",
    "count = data.shape[0]\n",
    "\n",
    "# No correlation\n",
    "no_corr_df = data[data['pearson'] == 0.0]\n",
    "no_correlation = no_corr_df.shape[0]\n",
    "\n",
    "# Weak positive (0 < r ≤ 0.2)\n",
    "weak_pos_df = data[(data['pearson'] > 0.0) & (data['pearson'] <= 0.2)]\n",
    "weak_pos = weak_pos_df.shape[0]\n",
    "\n",
    "# Weak negative (-0.2 ≤ r < 0)\n",
    "weak_neg_df = data[(data['pearson'] < 0.0) & (data['pearson'] >= -0.2)]\n",
    "weak_neg = weak_neg_df.shape[0]\n",
    "\n",
    "# Moderate positive (0.2 < r < 0.8)\n",
    "moderate_pos_df = data[(data['pearson'] > 0.2) & (data['pearson'] < 0.8)]\n",
    "moderate_pos = moderate_pos_df.shape[0]\n",
    "\n",
    "# Moderate negative (-0.8 < r < -0.2)\n",
    "moderate_neg_df = data[(data['pearson'] < -0.2) & (data['pearson'] > -0.8)]\n",
    "moderate_neg = moderate_neg_df.shape[0]\n",
    "\n",
    "# Strong positive (r ≥ 0.8)\n",
    "strong_pos_df = data[data['pearson'] >= 0.8]\n",
    "strong_pos = strong_pos_df.shape[0]\n",
    "\n",
    "# Strong negative (r ≤ -0.8)\n",
    "strong_neg_df = data[data['pearson'] <= -0.8]\n",
    "strong_neg = strong_neg_df.shape[0]\n",
    "\n",
    "# Print summary with direction\n",
    "print(f\"No correlation (r = 0.0): {no_correlation} ({no_correlation/count*100:.2f}%)\")\n",
    "print(f\"Weak positive (0 < r ≤ 0.2): {weak_pos} ({weak_pos/count*100:.2f}%)\")\n",
    "print(f\"Weak negative (-0.2 ≤ r < 0): {weak_neg} ({weak_neg/count*100:.2f}%)\")\n",
    "print(f\"Moderate positive (0.2 < r < 0.8): {moderate_pos} ({moderate_pos/count*100:.2f}%)\")\n",
    "print(f\"Moderate negative (-0.8 < r < -0.2): {moderate_neg} ({moderate_neg/count*100:.2f}%)\")\n",
    "print(f\"Strong positive (r ≥ 0.8): {strong_pos} ({strong_pos/count*100:.2f}%)\")\n",
    "print(f\"Strong negative (r ≤ -0.8): {strong_neg} ({strong_neg/count*100:.2f}%)\")\n",
    "\n",
    "# Supervisor’s key interest: total counts >0.2 and < -0.2\n",
    "pos_corr = data[data['pearson'] > 0.2].shape[0]\n",
    "neg_corr = data[data['pearson'] < -0.2].shape[0]\n",
    "\n",
    "print(f\"\\nCorrelations > 0.2: {pos_corr} ({pos_corr/count*100:.2f}%)\")\n",
    "print(f\"Correlations < -0.2: {neg_corr} ({neg_corr/count*100:.2f}%)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9246567c-3f6a-495b-8f43-fcad92c71abb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No correlation (r = 0.0): 5 (1.87%)\n",
      "Weak effect (0 < r ≤ 0.2): 175 (65.54%)\n",
      "Moderate effect (0.2 < r < 0.8): 80 (29.96%)\n",
      "Strong effect (r ≥ 0.8): 7 (2.62%)\n"
     ]
    }
   ],
   "source": [
    "# Effect size classifications\n",
    "count = data.shape[0]\n",
    "\n",
    "# Effect size classifications with proper grouping\n",
    "no_correlation = data[data['pearson'] == 0.0].shape[0]\n",
    "weak = data[(data['pearson'] > 0.0) & (data['pearson'] <= 0.2) | (data['pearson'] < 0.0) & (data['pearson'] >= -0.2)].shape[0]\n",
    "moderate = data[(data['pearson'] > 0.2) & (data['pearson'] < 0.8) | (data['pearson'] < -0.2) & (data['pearson'] >= -0.8)].shape[0]\n",
    "strong = data[(data['pearson'] >= 0.8) | (data['pearson'] <= -0.8)].shape[0]\n",
    "\n",
    "# Print out the classification results and proportions\n",
    "print(f\"No correlation (r = 0.0): {no_correlation} ({no_correlation/count*100:.2f}%)\")\n",
    "print(f\"Weak effect (0 < r ≤ 0.2): {weak} ({weak/count*100:.2f}%)\")\n",
    "print(f\"Moderate effect (0.2 > r < 0.8): {moderate} ({moderate/count*100:.2f}%)\")\n",
    "print(f\"Strong effect (r ≥ 0.8): {strong} ({strong/count*100:.2f}%)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4603a386-ab96-4ae1-b37b-a203e8ab8004",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_418132.1-345</th>\n",
       "      <td>-0.73</td>\n",
       "      <td>3.224524e-20</td>\n",
       "      <td>-0.72</td>\n",
       "      <td>9.821030e-20</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>3.224524e-20</td>\n",
       "      <td>1</td>\n",
       "      <td>6.622676e-19</td>\n",
       "      <td>cds-NP_418132.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418132.1-347</th>\n",
       "      <td>-0.73</td>\n",
       "      <td>3.224524e-20</td>\n",
       "      <td>-0.72</td>\n",
       "      <td>9.821030e-20</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>3.224524e-20</td>\n",
       "      <td>1</td>\n",
       "      <td>6.622676e-19</td>\n",
       "      <td>cds-NP_418132.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417082.1-201</th>\n",
       "      <td>0.82</td>\n",
       "      <td>7.630992e-29</td>\n",
       "      <td>0.37</td>\n",
       "      <td>3.777200e-05</td>\n",
       "      <td>0.82</td>\n",
       "      <td>7.630992e-29</td>\n",
       "      <td>1</td>\n",
       "      <td>6.460178e-27</td>\n",
       "      <td>cds-NP_417082.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418132.1-343</th>\n",
       "      <td>-0.73</td>\n",
       "      <td>3.224524e-20</td>\n",
       "      <td>-0.72</td>\n",
       "      <td>9.821030e-20</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>3.224524e-20</td>\n",
       "      <td>1</td>\n",
       "      <td>6.622676e-19</td>\n",
       "      <td>cds-NP_418132.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418089.1-1541</th>\n",
       "      <td>0.81</td>\n",
       "      <td>1.451725e-28</td>\n",
       "      <td>0.72</td>\n",
       "      <td>4.044563e-20</td>\n",
       "      <td>0.81</td>\n",
       "      <td>1.451725e-28</td>\n",
       "      <td>1</td>\n",
       "      <td>6.460178e-27</td>\n",
       "      <td>cds-NP_418089.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414752.1-2591</th>\n",
       "      <td>0.51</td>\n",
       "      <td>2.604887e-07</td>\n",
       "      <td>0.58</td>\n",
       "      <td>1.193219e-09</td>\n",
       "      <td>0.51</td>\n",
       "      <td>2.604887e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>2.575944e-06</td>\n",
       "      <td>cds-NP_414752.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418089.1-1527</th>\n",
       "      <td>0.81</td>\n",
       "      <td>1.451725e-28</td>\n",
       "      <td>0.72</td>\n",
       "      <td>4.044563e-20</td>\n",
       "      <td>0.81</td>\n",
       "      <td>1.451725e-28</td>\n",
       "      <td>1</td>\n",
       "      <td>6.460178e-27</td>\n",
       "      <td>cds-NP_418089.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418089.1-1533</th>\n",
       "      <td>0.81</td>\n",
       "      <td>1.451725e-28</td>\n",
       "      <td>0.72</td>\n",
       "      <td>4.044563e-20</td>\n",
       "      <td>0.81</td>\n",
       "      <td>1.451725e-28</td>\n",
       "      <td>1</td>\n",
       "      <td>6.460178e-27</td>\n",
       "      <td>cds-NP_418089.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_009518818.1-1883</th>\n",
       "      <td>-0.76</td>\n",
       "      <td>3.642711e-23</td>\n",
       "      <td>-0.53</td>\n",
       "      <td>1.033072e-09</td>\n",
       "      <td>-0.76</td>\n",
       "      <td>3.642711e-23</td>\n",
       "      <td>1</td>\n",
       "      <td>9.726039e-22</td>\n",
       "      <td>cds-YP_009518818.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_009518818.1-1897</th>\n",
       "      <td>-0.76</td>\n",
       "      <td>3.642711e-23</td>\n",
       "      <td>-0.53</td>\n",
       "      <td>1.033072e-09</td>\n",
       "      <td>-0.76</td>\n",
       "      <td>3.642711e-23</td>\n",
       "      <td>1</td>\n",
       "      <td>9.726039e-22</td>\n",
       "      <td>cds-YP_009518818.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418089.1-1530</th>\n",
       "      <td>0.81</td>\n",
       "      <td>1.451725e-28</td>\n",
       "      <td>0.72</td>\n",
       "      <td>4.044563e-20</td>\n",
       "      <td>0.81</td>\n",
       "      <td>1.451725e-28</td>\n",
       "      <td>1</td>\n",
       "      <td>6.460178e-27</td>\n",
       "      <td>cds-NP_418089.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418618.1-1616</th>\n",
       "      <td>0.49</td>\n",
       "      <td>2.417752e-08</td>\n",
       "      <td>0.33</td>\n",
       "      <td>2.495727e-04</td>\n",
       "      <td>0.49</td>\n",
       "      <td>2.417752e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>2.689749e-07</td>\n",
       "      <td>cds-NP_418618.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418492.1-369</th>\n",
       "      <td>0.41</td>\n",
       "      <td>5.449875e-06</td>\n",
       "      <td>0.47</td>\n",
       "      <td>1.396853e-07</td>\n",
       "      <td>0.41</td>\n",
       "      <td>5.449875e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>4.693925e-05</td>\n",
       "      <td>cds-NP_418492.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418618.1-1617</th>\n",
       "      <td>0.49</td>\n",
       "      <td>2.417752e-08</td>\n",
       "      <td>0.33</td>\n",
       "      <td>2.495727e-04</td>\n",
       "      <td>0.49</td>\n",
       "      <td>2.417752e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>2.689749e-07</td>\n",
       "      <td>cds-NP_418618.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418618.1-1615</th>\n",
       "      <td>0.49</td>\n",
       "      <td>2.417752e-08</td>\n",
       "      <td>0.33</td>\n",
       "      <td>2.495727e-04</td>\n",
       "      <td>0.49</td>\n",
       "      <td>2.417752e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>2.689749e-07</td>\n",
       "      <td>cds-NP_418618.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_009518818.1-1884</th>\n",
       "      <td>-0.76</td>\n",
       "      <td>3.642711e-23</td>\n",
       "      <td>-0.53</td>\n",
       "      <td>1.033072e-09</td>\n",
       "      <td>-0.76</td>\n",
       "      <td>3.642711e-23</td>\n",
       "      <td>1</td>\n",
       "      <td>9.726039e-22</td>\n",
       "      <td>cds-YP_009518818.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418514.1-2029</th>\n",
       "      <td>0.43</td>\n",
       "      <td>6.461417e-06</td>\n",
       "      <td>0.46</td>\n",
       "      <td>8.350655e-07</td>\n",
       "      <td>0.43</td>\n",
       "      <td>6.461417e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>5.391245e-05</td>\n",
       "      <td>cds-NP_418514.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418618.1-1618</th>\n",
       "      <td>0.49</td>\n",
       "      <td>2.417752e-08</td>\n",
       "      <td>0.33</td>\n",
       "      <td>2.495727e-04</td>\n",
       "      <td>0.49</td>\n",
       "      <td>2.417752e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>2.689749e-07</td>\n",
       "      <td>cds-NP_418618.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416077.4-3909</th>\n",
       "      <td>0.65</td>\n",
       "      <td>1.060612e-08</td>\n",
       "      <td>0.46</td>\n",
       "      <td>1.662010e-04</td>\n",
       "      <td>0.65</td>\n",
       "      <td>1.060612e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>1.415916e-07</td>\n",
       "      <td>cds-NP_416077.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418552.1-2265</th>\n",
       "      <td>-0.45</td>\n",
       "      <td>5.255163e-07</td>\n",
       "      <td>-0.45</td>\n",
       "      <td>5.337283e-07</td>\n",
       "      <td>-0.45</td>\n",
       "      <td>5.255163e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>5.011173e-06</td>\n",
       "      <td>cds-NP_418552.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417005.2-456</th>\n",
       "      <td>0.89</td>\n",
       "      <td>1.380083e-40</td>\n",
       "      <td>0.85</td>\n",
       "      <td>3.478063e-33</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1.380083e-40</td>\n",
       "      <td>1</td>\n",
       "      <td>3.684822e-38</td>\n",
       "      <td>cds-NP_417005.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415078.1-4012</th>\n",
       "      <td>0.51</td>\n",
       "      <td>1.242085e-07</td>\n",
       "      <td>0.53</td>\n",
       "      <td>2.994807e-08</td>\n",
       "      <td>0.51</td>\n",
       "      <td>1.242085e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>1.326546e-06</td>\n",
       "      <td>cds-NP_415078.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_009518818.1-1898</th>\n",
       "      <td>-0.76</td>\n",
       "      <td>3.642711e-23</td>\n",
       "      <td>-0.53</td>\n",
       "      <td>1.033072e-09</td>\n",
       "      <td>-0.76</td>\n",
       "      <td>3.642711e-23</td>\n",
       "      <td>1</td>\n",
       "      <td>9.726039e-22</td>\n",
       "      <td>cds-YP_009518818.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416543.1-2149</th>\n",
       "      <td>-0.54</td>\n",
       "      <td>2.165560e-06</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>8.449040e-04</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>2.165560e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>1.993809e-05</td>\n",
       "      <td>cds-NP_416543.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417599.1-3541</th>\n",
       "      <td>0.50</td>\n",
       "      <td>1.613982e-07</td>\n",
       "      <td>0.68</td>\n",
       "      <td>8.786553e-15</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1.613982e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>1.657435e-06</td>\n",
       "      <td>cds-NP_417599.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415918.1-2806</th>\n",
       "      <td>-0.57</td>\n",
       "      <td>2.664552e-06</td>\n",
       "      <td>0.13</td>\n",
       "      <td>3.400455e-01</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>2.664552e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>2.371451e-05</td>\n",
       "      <td>cds-NP_415918.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416476.1-1695</th>\n",
       "      <td>0.65</td>\n",
       "      <td>8.266797e-15</td>\n",
       "      <td>0.39</td>\n",
       "      <td>1.784522e-05</td>\n",
       "      <td>0.65</td>\n",
       "      <td>8.266797e-15</td>\n",
       "      <td>1</td>\n",
       "      <td>1.471490e-13</td>\n",
       "      <td>cds-NP_416476.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417235.1-4196</th>\n",
       "      <td>0.80</td>\n",
       "      <td>8.524843e-14</td>\n",
       "      <td>0.78</td>\n",
       "      <td>2.022357e-12</td>\n",
       "      <td>0.80</td>\n",
       "      <td>8.524843e-14</td>\n",
       "      <td>1</td>\n",
       "      <td>1.422583e-12</td>\n",
       "      <td>cds-NP_417235.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416720.1-0</th>\n",
       "      <td>0.54</td>\n",
       "      <td>6.665021e-10</td>\n",
       "      <td>0.18</td>\n",
       "      <td>5.703485e-02</td>\n",
       "      <td>0.54</td>\n",
       "      <td>6.665021e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>9.366108e-09</td>\n",
       "      <td>cds-NP_416720.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_009518815.1-100</th>\n",
       "      <td>-0.58</td>\n",
       "      <td>1.302597e-10</td>\n",
       "      <td>-0.41</td>\n",
       "      <td>1.232553e-05</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>1.302597e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>1.932185e-09</td>\n",
       "      <td>cds-YP_009518815.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414683.1-1552</th>\n",
       "      <td>0.57</td>\n",
       "      <td>7.514579e-11</td>\n",
       "      <td>0.64</td>\n",
       "      <td>5.637457e-14</td>\n",
       "      <td>0.57</td>\n",
       "      <td>7.514579e-11</td>\n",
       "      <td>1</td>\n",
       "      <td>1.180231e-09</td>\n",
       "      <td>cds-NP_414683.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417610.1-1604</th>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4.008327e-21</td>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.889370e-16</td>\n",
       "      <td>cds-NP_417610.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         pearson  pearson(p-value)  spearman  \\\n",
       "cds-NP_418132.1-345        -0.73      3.224524e-20     -0.72   \n",
       "cds-NP_418132.1-347        -0.73      3.224524e-20     -0.72   \n",
       "cds-NP_417082.1-201         0.82      7.630992e-29      0.37   \n",
       "cds-NP_418132.1-343        -0.73      3.224524e-20     -0.72   \n",
       "cds-NP_418089.1-1541        0.81      1.451725e-28      0.72   \n",
       "cds-NP_414752.1-2591        0.51      2.604887e-07      0.58   \n",
       "cds-NP_418089.1-1527        0.81      1.451725e-28      0.72   \n",
       "cds-NP_418089.1-1533        0.81      1.451725e-28      0.72   \n",
       "cds-YP_009518818.1-1883    -0.76      3.642711e-23     -0.53   \n",
       "cds-YP_009518818.1-1897    -0.76      3.642711e-23     -0.53   \n",
       "cds-NP_418089.1-1530        0.81      1.451725e-28      0.72   \n",
       "cds-NP_418618.1-1616        0.49      2.417752e-08      0.33   \n",
       "cds-NP_418492.1-369         0.41      5.449875e-06      0.47   \n",
       "cds-NP_418618.1-1617        0.49      2.417752e-08      0.33   \n",
       "cds-NP_418618.1-1615        0.49      2.417752e-08      0.33   \n",
       "cds-YP_009518818.1-1884    -0.76      3.642711e-23     -0.53   \n",
       "cds-NP_418514.1-2029        0.43      6.461417e-06      0.46   \n",
       "cds-NP_418618.1-1618        0.49      2.417752e-08      0.33   \n",
       "cds-NP_416077.4-3909        0.65      1.060612e-08      0.46   \n",
       "cds-NP_418552.1-2265       -0.45      5.255163e-07     -0.45   \n",
       "cds-NP_417005.2-456         0.89      1.380083e-40      0.85   \n",
       "cds-NP_415078.1-4012        0.51      1.242085e-07      0.53   \n",
       "cds-YP_009518818.1-1898    -0.76      3.642711e-23     -0.53   \n",
       "cds-NP_416543.1-2149       -0.54      2.165560e-06     -0.39   \n",
       "cds-NP_417599.1-3541        0.50      1.613982e-07      0.68   \n",
       "cds-NP_415918.1-2806       -0.57      2.664552e-06      0.13   \n",
       "cds-NP_416476.1-1695        0.65      8.266797e-15      0.39   \n",
       "cds-NP_417235.1-4196        0.80      8.524843e-14      0.78   \n",
       "cds-NP_416720.1-0           0.54      6.665021e-10      0.18   \n",
       "cds-YP_009518815.1-100     -0.58      1.302597e-10     -0.41   \n",
       "cds-NP_414683.1-1552        0.57      7.514579e-11      0.64   \n",
       "cds-NP_417610.1-1604        0.69      9.906809e-18      0.74   \n",
       "\n",
       "                         spearman(p-value)  r_value  regress(p-value)  \\\n",
       "cds-NP_418132.1-345           9.821030e-20    -0.73      3.224524e-20   \n",
       "cds-NP_418132.1-347           9.821030e-20    -0.73      3.224524e-20   \n",
       "cds-NP_417082.1-201           3.777200e-05     0.82      7.630992e-29   \n",
       "cds-NP_418132.1-343           9.821030e-20    -0.73      3.224524e-20   \n",
       "cds-NP_418089.1-1541          4.044563e-20     0.81      1.451725e-28   \n",
       "cds-NP_414752.1-2591          1.193219e-09     0.51      2.604887e-07   \n",
       "cds-NP_418089.1-1527          4.044563e-20     0.81      1.451725e-28   \n",
       "cds-NP_418089.1-1533          4.044563e-20     0.81      1.451725e-28   \n",
       "cds-YP_009518818.1-1883       1.033072e-09    -0.76      3.642711e-23   \n",
       "cds-YP_009518818.1-1897       1.033072e-09    -0.76      3.642711e-23   \n",
       "cds-NP_418089.1-1530          4.044563e-20     0.81      1.451725e-28   \n",
       "cds-NP_418618.1-1616          2.495727e-04     0.49      2.417752e-08   \n",
       "cds-NP_418492.1-369           1.396853e-07     0.41      5.449875e-06   \n",
       "cds-NP_418618.1-1617          2.495727e-04     0.49      2.417752e-08   \n",
       "cds-NP_418618.1-1615          2.495727e-04     0.49      2.417752e-08   \n",
       "cds-YP_009518818.1-1884       1.033072e-09    -0.76      3.642711e-23   \n",
       "cds-NP_418514.1-2029          8.350655e-07     0.43      6.461417e-06   \n",
       "cds-NP_418618.1-1618          2.495727e-04     0.49      2.417752e-08   \n",
       "cds-NP_416077.4-3909          1.662010e-04     0.65      1.060612e-08   \n",
       "cds-NP_418552.1-2265          5.337283e-07    -0.45      5.255163e-07   \n",
       "cds-NP_417005.2-456           3.478063e-33     0.89      1.380083e-40   \n",
       "cds-NP_415078.1-4012          2.994807e-08     0.51      1.242085e-07   \n",
       "cds-YP_009518818.1-1898       1.033072e-09    -0.76      3.642711e-23   \n",
       "cds-NP_416543.1-2149          8.449040e-04    -0.54      2.165560e-06   \n",
       "cds-NP_417599.1-3541          8.786553e-15     0.50      1.613982e-07   \n",
       "cds-NP_415918.1-2806          3.400455e-01    -0.57      2.664552e-06   \n",
       "cds-NP_416476.1-1695          1.784522e-05     0.65      8.266797e-15   \n",
       "cds-NP_417235.1-4196          2.022357e-12     0.80      8.524843e-14   \n",
       "cds-NP_416720.1-0             5.703485e-02     0.54      6.665021e-10   \n",
       "cds-YP_009518815.1-100        1.232553e-05    -0.58      1.302597e-10   \n",
       "cds-NP_414683.1-1552          5.637457e-14     0.57      7.514579e-11   \n",
       "cds-NP_417610.1-1604          4.008327e-21     0.69      9.906809e-18   \n",
       "\n",
       "                         association  adjusted_p_values                gene  \n",
       "cds-NP_418132.1-345                1       6.622676e-19     cds-NP_418132.1  \n",
       "cds-NP_418132.1-347                1       6.622676e-19     cds-NP_418132.1  \n",
       "cds-NP_417082.1-201                1       6.460178e-27     cds-NP_417082.1  \n",
       "cds-NP_418132.1-343                1       6.622676e-19     cds-NP_418132.1  \n",
       "cds-NP_418089.1-1541               1       6.460178e-27     cds-NP_418089.1  \n",
       "cds-NP_414752.1-2591               1       2.575944e-06     cds-NP_414752.1  \n",
       "cds-NP_418089.1-1527               1       6.460178e-27     cds-NP_418089.1  \n",
       "cds-NP_418089.1-1533               1       6.460178e-27     cds-NP_418089.1  \n",
       "cds-YP_009518818.1-1883            1       9.726039e-22  cds-YP_009518818.1  \n",
       "cds-YP_009518818.1-1897            1       9.726039e-22  cds-YP_009518818.1  \n",
       "cds-NP_418089.1-1530               1       6.460178e-27     cds-NP_418089.1  \n",
       "cds-NP_418618.1-1616               1       2.689749e-07     cds-NP_418618.1  \n",
       "cds-NP_418492.1-369                1       4.693925e-05     cds-NP_418492.1  \n",
       "cds-NP_418618.1-1617               1       2.689749e-07     cds-NP_418618.1  \n",
       "cds-NP_418618.1-1615               1       2.689749e-07     cds-NP_418618.1  \n",
       "cds-YP_009518818.1-1884            1       9.726039e-22  cds-YP_009518818.1  \n",
       "cds-NP_418514.1-2029               1       5.391245e-05     cds-NP_418514.1  \n",
       "cds-NP_418618.1-1618               1       2.689749e-07     cds-NP_418618.1  \n",
       "cds-NP_416077.4-3909               1       1.415916e-07     cds-NP_416077.4  \n",
       "cds-NP_418552.1-2265               1       5.011173e-06     cds-NP_418552.1  \n",
       "cds-NP_417005.2-456                1       3.684822e-38     cds-NP_417005.2  \n",
       "cds-NP_415078.1-4012               1       1.326546e-06     cds-NP_415078.1  \n",
       "cds-YP_009518818.1-1898            1       9.726039e-22  cds-YP_009518818.1  \n",
       "cds-NP_416543.1-2149               1       1.993809e-05     cds-NP_416543.1  \n",
       "cds-NP_417599.1-3541               1       1.657435e-06     cds-NP_417599.1  \n",
       "cds-NP_415918.1-2806               1       2.371451e-05     cds-NP_415918.1  \n",
       "cds-NP_416476.1-1695               1       1.471490e-13     cds-NP_416476.1  \n",
       "cds-NP_417235.1-4196               1       1.422583e-12     cds-NP_417235.1  \n",
       "cds-NP_416720.1-0                  1       9.366108e-09     cds-NP_416720.1  \n",
       "cds-YP_009518815.1-100             1       1.932185e-09  cds-YP_009518815.1  \n",
       "cds-NP_414683.1-1552               1       1.180231e-09     cds-NP_414683.1  \n",
       "cds-NP_417610.1-1604               1       1.889370e-16     cds-NP_417610.1  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filtered = data[data['adjusted_p_values'] < 0.0001]\n",
    "data_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "872f96ea-67f1-4630-8350-2f53f183a674",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "ZeroDivisionError",
     "evalue": "float division by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [22]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m non_sig \u001b[38;5;241m=\u001b[39m prediction[prediction[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124massociation\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Perform multiple test correlation using Benjamini-Hochberg (FDR) correction\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m non_sig[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124madjusted_p_values\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mmultipletests\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnon_sig\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpearson(p-value)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfdr_bh\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m      6\u001b[0m non_sig\u001b[38;5;241m.\u001b[39mto_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnon-associated-correlation-panfeed230.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m, sep \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      7\u001b[0m non_sig\n",
      "File \u001b[0;32m/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/statsmodels/stats/multitest.py:147\u001b[0m, in \u001b[0;36mmultipletests\u001b[0;34m(pvals, alpha, method, is_sorted, returnsorted)\u001b[0m\n\u001b[1;32m    144\u001b[0m     pvals \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mtake(pvals, sortind)\n\u001b[1;32m    146\u001b[0m ntests \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(pvals)\n\u001b[0;32m--> 147\u001b[0m alphacSidak \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;241m-\u001b[39m np\u001b[38;5;241m.\u001b[39mpower((\u001b[38;5;241m1.\u001b[39m \u001b[38;5;241m-\u001b[39m alphaf), \u001b[38;5;241;43m1.\u001b[39;49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[43mntests\u001b[49m)\n\u001b[1;32m    148\u001b[0m alphacBonf \u001b[38;5;241m=\u001b[39m alphaf \u001b[38;5;241m/\u001b[39m \u001b[38;5;28mfloat\u001b[39m(ntests)\n\u001b[1;32m    149\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m method\u001b[38;5;241m.\u001b[39mlower() \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbonf\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbonferroni\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n",
      "\u001b[0;31mZeroDivisionError\u001b[0m: float division by zero"
     ]
    }
   ],
   "source": [
    "non_sig = prediction[prediction['association'] == 0]\n",
    "\n",
    "\n",
    "# Perform multiple test correlation using Benjamini-Hochberg (FDR) correction\n",
    "non_sig['adjusted_p_values'] = multipletests(non_sig['pearson(p-value)'], method='fdr_bh')[1]\n",
    "non_sig.to_csv('non-associated-correlation-panfeed230.tsv', sep = '\\t')\n",
    "non_sig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e7e6bf8d-e9e6-4d20-b566-85bf5f36f60d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For threshold 0.001, number of significant correlations is: 426 (9.31%)\n",
      "For threshold 0.002, number of significant correlations is: 451 (9.85%)\n",
      "For threshold 0.003, number of significant correlations is: 452 (9.88%)\n",
      "For threshold 0.004, number of significant correlations is: 464 (10.14%)\n",
      "For threshold 0.005, number of significant correlations is: 464 (10.14%)\n",
      "For threshold 0.0001, number of significant correlations is: 370 (8.08%)\n"
     ]
    }
   ],
   "source": [
    "thresholds = [0.001,0.002, 0.003, 0.004, 0.005, 0.0001]  \n",
    "\n",
    "total_var = len(non_sig)\n",
    "\n",
    "for threshold in thresholds:\n",
    "    significant_correlations = non_sig[(non_sig['pearson'] > 0) & (non_sig['adjusted_p_values'] <= threshold)]\n",
    "    num_significant = len(significant_correlations)\n",
    "    percentage = (num_significant / total_var) * 100\n",
    "    print(f\"For threshold {threshold}, number of significant correlations is: {num_significant} ({percentage:.2f}%)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8a239bf5-c8c2-4ac9-bcb5-f2bc6690e4c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEWCAYAAACJ0YulAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAhiUlEQVR4nO3de5gdVZnv8e+PBAJBhMQk2ORCACOc4MPNJiKgggEJCASPZgxeTgbBiAcvODoScAZRzAjjGQ/jmRM1KmMAJQa5RUQlBgOKQGiQWxJiggFyI2kQCDcjie/8UatrKt27u3dfau/u5Pd5nv3sqlVrVb1VvXu/u26rFBGYmZkB7FTvAMzMrO9wUjAzs5yTgpmZ5ZwUzMws56RgZmY5JwUzM8s5KZh1kaRfSJpWo2VdIumaGi3rO5L+uRbLsr7LScG6TdITkjZI2r1Qdo6kRXUMq3QRcXJEzKmmbtpGJ5QdU2+IiHMj4tKezkfSWEkhaWBvxGW15aRgPTUQ+Gy9g7CekTSg3jFY3+CkYD31DeALkvaqNFHS0ZLuk/RCej+6MG2RpEsl3SXpRUm3SRrW3oIknSVpWar7J0mfKEwbJukWSc9L+rOk30raKU27QNLa1G65pImpfJCkKyStS68rJA0qzHOypAclbZL0uKRJhbjPScMHSLpd0rOSnpH0o5ZtIelqYAzwM0kvSfpiKj9K0u9TrA9JOq6wzP0k3ZFiXQB0tD2WSTq1MD4wxXBEGr9O0tNp298p6eBC3R9K+rakWyW9DByfyr6Wpg9J27NZ0nNpeFSVf7s70/vzab3fLulNab1eSDH+pL31sjqLCL/86tYLeAI4AbgB+FoqOwdYlIaHAs8BHyXbozgzjb8hTV8EPA68GdgtjV/WwfLeCxwACHgX8ApwRJr2deA7wM7p9Y5U70BgNbBPqjcWOCANfxW4BxgBDAd+D1yapk0AXgBOJPvxNBI4qBD3OWn4TanOoDSPO4ErWm+jwvhI4FnglDTfE9P48DT9buCbaX7vBF4Ermlne1wM/KjV9nmsMP4xYI80ryuABwvTfpjW75gUx66prOXv+Abg/cDgNI/rgJsK7dv926VtHMDAQv1rgS8VlnVsvT+/flV+eU/BesPFwKclDW9V/l5gRURcHRFbIuJa4DHgtEKd/4yIP0bEq8A84LD2FhIRP4+IxyNzB3Ab2Zc/wGtAA7BvRLwWEb+N7NtoK9mX4nhJO0fEExHxeGrzYeCrEbExIpqBr5AlMICzgSsjYkFE/C0i1kbEYxViWpnqbE7z+CZZwmrPR4BbI+LWNN8FQBNwiqQxwJHAP6f53Qn8rIN5/Rg4XdLgNP6hVNYS25UR8WJEbAYuAQ6VtGeh/c0RcVeK4y+t1uvZiLg+Il6JiBeBmRXWq+q/HdnfZ1+y5PyXiPhdB3WtjpwUrMci4lHgFmBGq0n7AE+2KnuS7Ndyi6cLw68Ar4P8SpiX0uuiVHaypHvS4aHnyX5ttxyy+AawErgtHVqakWJbCZxP9qW4UdJcSfu0E9+TqQxgNNkv4Q5JGpHmuVbSJuAaOjjkQ/bFOCUdOno+rcexZAltH+C5iHi5VUwVpXVbBpyWEsPppKQgaYCky9Jhr01keyy0im11B+s1WNJ3JT2Z2t8J7KVtzz1U/Nu144tke26LJS2R9LEO6lodOSlYb/ky8HG2/cJfR/YlWDQGWNvZzCK7EuZ16fUv6Vj/9cD/AfaOiL2AW8m+aEi/iD8fEfuT7Yn8Q8u5g4j4cUQcm2IJ4PJ24huTyiD7wjygivX+eprnIRHxerI9ARVXpVX91cDVEbFX4bV7RFwGrAeGqHA1V4qpI9eSHZabDCxNiQKyvYbJZIf39iQ7pEMnsRV9nuzQ29vSer2zQvv2tJlvRDwdER+PiH2ATwCzJL2pinlZjTkpWK9IX0Y/AT5TKL4VeLOkD6WToB8ExpPtVXTVLmSHgZqBLZJOBt7TMlHSqelkpoBNZIeNtko6UNK7U1L5C/BqmgbZF+o/SRqeTpJeTPZLH+AHwFmSJkraSdJISQdViGsP4CWyk6ojgX9sNX0DsH9h/BqyX/YnpV/zu0o6TtKoiHiS7FDSVyTtIulYtj3UVsnctB0+SeHQUYprM9n5isHAv3Qyn0rr9Wpar6FkSb9azcDfKKy3pCmFE9XPkSWOrRXaWp05KVhv+iqQ/8qNiGeBU8l+dT5Ldgjh1Ih4pqszTse1P0N27Po5sl/C8wtVxgG/JvuCvhuYFRGLyBLJZcAzZIc7RgAXpTZfI/sSfhh4BHgglRERi4GzgP9LdkL2Dtru9UB2HuKIVOfnZCfdi75Olniel/SFiFhN9gv+IrIvz9VkiaTlf/FDwNuAP5N9EV/VyXZZn9b3aLKk3OIqskNPa4GlZCfUu+IKshPIz6S2v6y2YUS8QnYO4q603keRnSu5V9JLZH+3z0bEqi7GZDWg7FycmZmZ9xTMzKzAScHMzHJOCmZmlnNSMDOzXL/uxXDYsGExduzYeodhZtav3H///c9EROseCIB+nhTGjh1LU1NTvcMwM+tXJLV7p3yph48kfS7d0v6opGvTjTpDJS2QtCK9DynUv1DSSmU9WZ5UZmxmZtZWaUkh3d35GaAxIt4CDACmkvWPszAixgEL0ziSxqfpBwOTyG6Ddx/vZmY1VPaJ5oHAbsqewDSYrF+ZyUDLU6vmAGek4cnA3NQ75Cqyzs0mlByfmZkVlJYUImItWedlT5F19PVCRNxG1pnZ+lRnPVm3A5B1pFbstXEN23auBoCk6ZKaJDU1NzeXFb6Z2Q6pzMNHQ8h+/e9H1iXw7pI+0lGTCmWVelucHRGNEdE4fHjFk+dmZtZNZR4+OgFYFRHNEfEaWUdhRwMbJDUApPeNqf4asj7sW4ziv7sxNjOzGigzKTwFHJUe1iFgItkDQeYD01KdacDNaXg+MFXZc3P3I+v1cnGJ8ZmZWSul3acQEfdK+ilZd8RbgD8As8mezjRP0tlkiWNKqr9E0jyybn63AOdFhPtbNzOroX7ddXZjY2P45jUzs66RdH9ENFaa5r6PbLvWMGoMkrr9ahjV2dMwzbYv/bqbC7POPL12Nfte0J2nf2aevPzUXozGrO/znoKZmeWcFMzMLOekYGZmOScFMzPLOSmYmVnOScHMzHJOCmZmlnNSMDOznJOCmZnlnBTMzCznpGBmZjknBTMzyzkpmJlZzknBzMxyTgpmZpYrLSlIOlDSg4XXJknnSxoqaYGkFel9SKHNhZJWSlou6aSyYjMzs8pKSwoRsTwiDouIw4C3Aq8ANwIzgIURMQ5YmMaRNB6YChwMTAJmSRpQVnxmZtZWrQ4fTQQej4gngcnAnFQ+BzgjDU8G5kbE5ohYBawEJtQoPjMzo3ZJYSpwbRreOyLWA6T3Eal8JLC60GZNKtuGpOmSmiQ1NTc3lxiymdmOp/SkIGkX4HTgus6qViiLNgURsyOiMSIahw8f3hshmrVvwM5I6tarYdSYekdv1mUDa7CMk4EHImJDGt8gqSEi1ktqADam8jXA6EK7UcC6GsRn1r6tr7HvBbd0q+mTl5/ay8GYla8Wh4/O5L8PHQHMB6al4WnAzYXyqZIGSdoPGAcsrkF8ZmaWlLqnIGkwcCLwiULxZcA8SWcDTwFTACJiiaR5wFJgC3BeRGwtMz4zM9tWqUkhIl4B3tCq7Fmyq5Eq1Z8JzCwzJjMza5/vaDYzs5yTgpmZ5ZwUzMws56RgZmY5JwUzM8s5KZiZWc5JwczMck4KZmaWc1IwM7Ock4JZWdzDqvVDtegl1WzH5B5WrR/ynoKZmeWcFMzMLOekYGZmOScFMzPLOSmYmVnOScHMzHJOCmZmlis1KUjaS9JPJT0maZmkt0saKmmBpBXpfUih/oWSVkpaLumkMmMzM7O2yt5T+HfglxFxEHAosAyYASyMiHHAwjSOpPHAVOBgYBIwS9KAkuMzM7OC0pKCpNcD7wR+ABARf42I54HJwJxUbQ5wRhqeDMyNiM0RsQpYCUwoKz4zM2urzD2F/YFm4D8l/UHS9yXtDuwdEesB0vuIVH8ksLrQfk0q24ak6ZKaJDU1NzeXGL6Z2Y6nzKQwEDgC+HZEHA68TDpU1A5VKIs2BRGzI6IxIhqHDx/eO5GamRlQblJYA6yJiHvT+E/JksQGSQ0A6X1jof7oQvtRwLoS4zMzs1ZKSwoR8TSwWtKBqWgisBSYD0xLZdOAm9PwfGCqpEGS9gPGAYvLis/MzNoqu+vsTwM/krQL8CfgLLJENE/S2cBTwBSAiFgiaR5Z4tgCnBcRW0uOz8zMCkpNChHxINBYYdLEdurPBGaWGZOZmbXPdzSbmVnOScHMzHJOCmZmlnNSMDOznJOCmZnlnBSsz2sYNQZJ3XqZWdeUfZ+CWY89vXY1+15wS7faPnn5qb0cjdn2zXsKZmaWc1IwM7Ock4KZmeWcFMzMLOekYGZmOScFMzPLOSmYmVnOScGsLxqwc7dv2GsYNabe0Vs/5pvXzPqira/5hj2rC+8pmJlZzknBzMxypSYFSU9IekTSg5KaUtlQSQskrUjvQwr1L5S0UtJySSeVGZuZmbVViz2F4yPisIhoeVbzDGBhRIwDFqZxJI0HpgIHA5OAWZIG1CA+MzNL6nH4aDIwJw3PAc4olM+NiM0RsQpYCUyofXhmZjuuspNCALdJul/S9FS2d0SsB0jvI1L5SGB1oe2aVLYNSdMlNUlqam5uLjF0M7MdT1VJQdJbujn/YyLiCOBk4DxJ7+xoMRXKok1BxOyIaIyIxuHDh3czLDMzq6TaPYXvSFos6X9L2qvamUfEuvS+EbiR7HDQBkkNAOl9Y6q+BhhdaD4KWFftsszMrOeqSgoRcSzwYbIv7SZJP5Z0YkdtJO0uaY+WYeA9wKPAfGBaqjYNuDkNzwemShokaT9gHLC4i+tjZmY9UPUdzRGxQtI/AU3At4DDlT0E96KIuKFCk72BG9NzcgcCP46IX0q6D5gn6WzgKWBKmv8SSfOApcAW4LyI2NqDdTMzsy6qKilIOgQ4C3gvsAA4LSIekLQPcDfQJilExJ+AQyuUPwtMrLSciJgJzKw6ejMz61XV7in8B/A9sr2CV1sKI2Jd2nswM7PtQLVJ4RTg1ZbDOZJ2AnaNiFci4urSojMzs5qq9uqjXwO7FcYHpzIzM9uOVJsUdo2Il1pG0vDgckIyM7N6qTYpvCzpiJYRSW8FXu2gvpmZ9UPVnlM4H7hOUsvNZA3AB0uJyMzM6qaqpBAR90k6CDiQrDuKxyLitVIjMzOzmuvK4ziPBMamNodLIiKuKiUqMzOri2pvXrsaOAB4EGi5yzgAJwUzs+1ItXsKjcD4iGjTa6mZmW0/qr366FHgjWUGYmZm9VftnsIwYKmkxcDmlsKIOL2UqMzMrC6qTQqXlBmEmZn1DdVeknqHpH2BcRHxa0mDgQHlhmZmZrVW7eM4Pw78FPhuKhoJ3FRSTGZmVifVnmg+DzgG2ATZA3eAEWUFZWZm9VFtUtgcEX9tGZE0kOw+BTMz245UmxTukHQRsFt6NvN1wM+qaShpgKQ/SLoljQ+VtEDSivQ+pFD3QkkrJS2XdFJXV8bMzHqm2qQwA2gGHgE+AdwKVPvEtc8Cy1rNa2FEjAMWpnEkjQemAgcDk4BZknwy28yshqpKChHxt4j4XkRMiYgPpOFODx9JGkX2XOfvF4onA3PS8BzgjEL53IjYHBGrgJXAhCrXw8zMekG1fR+tosI5hIjYv5OmVwBfBPYolO0dEetT+/WSWk5YjwTuKdRbk8paxzIdmA4wZsyYasI3M7MqdaXvoxa7AlOAoR01kHQqsDEi7pd0XBXLUIWySoloNjAboLGx0Se7zcx6UbU3rz3bqugKSb8DLu6g2THA6ZJOIUskr5d0DbBBUkPaS2gANqb6a4DRhfajgHWYmVnNVHvz2hGFV6Okc9n2kFAbEXFhRIyKiLFkJ5Bvj4iPAPOBaanaNODmNDwfmCppkKT9gHHA4q6vkpmZdVe1h4/+rTC8BXgC+LtuLvMyYJ6ks4GnyA5FERFLJM0DlqZlnBcRW9ufjZmZ9bZqDx8d35OFRMQiYFEafhaY2E69mcDMnizLzMy6r9qrj/6ho+kR8c3eCcfMzOqpK1cfHUl23B/gNOBOYHUZQZmZWX105SE7R0TEiwCSLgGui4hzygrMzMxqr9puLsYAfy2M/xUY2+vRmJlZXVW7p3A1sFjSjWQ3lL0PuKq0qMzMrC6qvfpopqRfAO9IRWdFxB/KC8vMzOqh2sNHAIOBTRHx78CadIOZmZltR6q9o/nLwAXAhaloZ+CasoIyM7P6qHZP4X3A6cDLABGxjk66uTAzs/6n2qTw1/T8hACQtHt5IZmZWb1UmxTmSfousJekjwO/Br5XXlhmZlYPnV59JEnAT4CDgE3AgcDFEbGg5NjMzKzGOk0KERGSboqItwJOBGZm27FqDx/dI+nIUiMxM7O6q/aO5uOBcyU9QXYFksh2Ig4pKzAzM6u9DpOCpDER8RRwco3iMTOzOupsT+Emst5Rn5R0fUS8vwYxmZlZnXR2TkGF4f3LDMTMzOqvs6QQ7Qx3StKukhZLekjSEklfSeVDJS2QtCK9Dym0uVDSSknLJZ3UleWZmVnPdZYUDpW0SdKLwCFpeJOkFyVt6qTtZuDdEXEocBgwSdJRwAxgYUSMAxamcSSNB6YCBwOTgFmSBnR7zczMrMs6TAoRMSAiXh8Re0TEwDTcMv76TtpGRLyURndOrwAmA3NS+RzgjDQ8GZgbEZsjYhWwEpjQvdUyM7Pu6ErX2V0maYCkB4GNwIKIuBfYOyLWA6T3Ean6SLZ95vOaVNZ6ntMlNUlqam5uLjN8M7MdTqlJISK2RsRhwChggqS3dFBdFcranMeIiNkR0RgRjcOHD++lSM3MDEpOCi0i4nlgEdm5gg2SGgDS+8ZUbQ0wutBsFLCuFvGZmVmmtKQgabikvdLwbsAJwGPAfGBaqjYNuDkNzwemShqUnuo2DlhcVnxmZtZWtd1cdEcDMCddQbQTMC8ibpF0N1lX3GcDTwFTACJiiaR5wFJgC3BeRGwtMT4zM2ultKQQEQ8Dh1cofxaY2E6bmcDMsmIyM7OO1eScgpnV0ICdkdStV8OoMfWO3uqszMNHZlYPW19j3wtu6VbTJy8/tZeDsf7GewpmZpZzUjAzs5yTgpmZ5ZwUzMws56RgZmY5JwUzM8s5KZiZWc5JwczMck4KZmaWc1IwM7Ock4KZmeWcFMzMLOekYGZmOScFMzPLOSmYmVmuzGc0j5b0G0nLJC2R9NlUPlTSAkkr0vuQQpsLJa2UtFzSSWXFZmZmlZW5p7AF+HxE/A/gKOA8SeOBGcDCiBgHLEzjpGlTgYOBScCs9HxnMzOrkdKSQkSsj4gH0vCLwDJgJDAZmJOqzQHOSMOTgbkRsTkiVgErgQllxWdmZm3V5JyCpLHA4cC9wN4RsR6yxAGMSNVGAqsLzdakMjMzq5HSk4Kk1wHXA+dHxKaOqlYoiwrzmy6pSVJTc3Nzb4W5Q2gYNabbD3T3Q93NdgwDy5y5pJ3JEsKPIuKGVLxBUkNErJfUAGxM5WuA0YXmo4B1recZEbOB2QCNjY1tkoa17+m1q7v9QHfwQ93NdgRlXn0k4AfAsoj4ZmHSfGBaGp4G3FwonyppkKT9gHHA4rLiMzOztsrcUzgG+CjwiKQHU9lFwGXAPElnA08BUwAiYomkecBSsiuXzouIrSXGZ2ZmrZSWFCLid1Q+TwAwsZ02M4GZZcVkZmYd8x3NZmaWc1IwM7Ock4KZmeWcFKwmenKPhJnVTqn3KZi16Mk9Er4/wqx2vKdgZmY5JwUzM8s5KZiZWc5JwczMck4KZmaWc1Iwsz6hJ5ctu1v33uNLUs2sT/Bly32D9xTMzCznpGBmZjknBTMzyzkpmJlZzknBzMxyTgpmZpYrLSlIulLSRkmPFsqGSlogaUV6H1KYdqGklZKWSzqprLjMzKx9Ze4p/BCY1KpsBrAwIsYBC9M4ksYDU4GDU5tZkgaUGJuZlcDPzej/Srt5LSLulDS2VfFk4Lg0PAdYBFyQyudGxGZglaSVwATg7rLiM7Pe5xvQ+r9an1PYOyLWA6T3Eal8JLC6UG9NKmtD0nRJTZKampubSw3WzKxMfbFrj77SzUWlfceoVDEiZgOzARobGyvWMTPrD/rinlWt9xQ2SGoASO8bU/kaYHSh3ihgXY1jMzPb4dU6KcwHpqXhacDNhfKpkgZJ2g8YByyucWxmZju80g4fSbqW7KTyMElrgC8DlwHzJJ0NPAVMAYiIJZLmAUuBLcB5EbG1rNjMzKyyMq8+OrOdSRPbqT8TmFlWPGZm1rm+cqLZ+oMBO/t68u2d/8Y7PCeFOmgYNYan167uvGIFbxw5mvVrnurliKq09bU+d6WE9bIe/I3Bf+ftgZNCHfTFy9DMzMAd4pmZWYGTgpmZ5ZwUzMy6qSfdVPTVE/o+p2Bm1k09OT8IffMcofcUzMws56RgZmY5JwUzM8s5KZiZWc5JwczMcr76qL9x3zRmViInhW7qSf9FPeL+h8ysRE4K3eT+i8xse7RDn1Poyd2IZmbbox16T8G/9s3MtrVD7ymYmdm2+lxSkDRJ0nJJKyXNqHc8ZmY7kj6VFCQNAP4/cDIwHjhT0vj6RmVmtuPoU0kBmACsjIg/RcRfgbnA5DrHZGbbMV9wsi1FRL1jyEn6ADApIs5J4x8F3hYRnyrUmQ5MT6MHAstrHmhmGPBMnZbdHY63PP0pVnC8ZesP8e4bEcMrTehrVx9VSr3bZK2ImA3Mrk047ZPUFBGN9Y6jWo63PP0pVnC8Zetv8bbW1w4frQFGF8ZHAevqFIuZ2Q6nryWF+4BxkvaTtAswFZhf55jMzHYYferwUURskfQp4FfAAODKiFhS57DaU/dDWF3keMvTn2IFx1u2/hbvNvrUiWYzM6uvvnb4yMzM6shJwczMck4KHZA0VNICSSvS+5AKdQ6U9GDhtUnS+WnaJZLWFqadUu94U70nJD2SYmrqavtaxSpptKTfSFomaYmkzxam1WTbdtbtijLfStMflnREtW3rFO+HU5wPS/q9pEML0yp+LuoY63GSXij8jS+utm2d4v3HQqyPStoqaWiaVtNt2yMR4Vc7L+BfgRlpeAZweSf1BwBPk90YAnAJ8IW+Fi/wBDCsp+tbdqxAA3BEGt4D+CMwvlbbNv09Hwf2B3YBHmpZfqHOKcAvyO6xOQq4t9q2dYr3aGBIGj65Jd6OPhd1jPU44JbutK1HvK3qnwbcXo9t29OX9xQ6NhmYk4bnAGd0Un8i8HhEPFlmUB3oary93b5XlxUR6yPigTT8IrAMGFliTK1V0+3KZOCqyNwD7CWpocq2NY83In4fEc+l0XvI7gWqh55snz65bVs5E7i25JhK4aTQsb0jYj1kX1DAiE7qT6XtB+FTaVf9yjIPxyTVxhvAbZLuT92GdLV9LWMFQNJY4HDg3kJx2dt2JFB85uoa2ial9upU07a3dXWZZ5Pt5bRo73NRhmpjfbukhyT9QtLBXWzbm6pepqTBwCTg+kJxLbdtj/Sp+xTqQdKvgTdWmPSlLs5nF+B04MJC8beBS8k+EJcC/wZ8rHuR5svpjXiPiYh1kkYACyQ9FhF39iSuSnpx276O7B/s/IjYlIp7fdtWWnSFstbXcLdXp5q2va3qZUo6niwpHFsorsnnoiWECmWtY32A7FDsS+mc0U3AuCrb9rauLPM04K6I+HOhrJbbtkd2+KQQESe0N03SBkkNEbE+HRLY2MGsTgYeiIgNhXnnw5K+B3TvMW+9HG9ErEvvGyXdSLZrfCfQlfWtSaySdiZLCD+KiBsK8+71bVtBNd2utFdnlyra9raquomRdAjwfeDkiHi2pbyDz0VdYi38ACAibpU0S9KwatqWoCvLbHPEoMbbtkd8+Khj84FpaXgacHMHddscQ0xfdi3eBzzaq9G11Wm8knaXtEfLMPCeQlxdWd9axCrgB8CyiPhmq2m12LbVdLsyH/hf6Sqko4AX0uGwenTZ0ukyJY0BbgA+GhF/LJR39LmoV6xvTJ8BJE0g+756tpq29Yg3xbkn8C4Kn+c6bNueqfeZ7r78At4ALARWpPehqXwf4NZCvcFkH9Y9W7W/GngEeJjsA9RQ73jJrp54KL2WAF/qrH0dYz2WbBf9YeDB9DqlltuW7OqiP5JdefKlVHYucG4aFtmDoR5P8TR21LYGn9nO4v0+8FxhezZ19rmoY6yfSrE8RHZS/Oi+vG3T+N8Dc1u1q/m27cnL3VyYmVnOh4/MzCznpGBmZjknBTMzyzkpmJlZzknBzMxyTgrWL6Rr1udKelzSUkm3SnpzCcsZK6nDa8hTnQ8Vxhslfau3Y6lG6n1zWD2WbdsnJwXr89INTDcCiyLigIgYD1wE7F1l+wEdjXfDWCBPChHRFBGf6eE8zfoEJwXrD44HXouI77QURMSDEfHbdCfxN1L/9Y9I+iDkffH/RtKPgUcqjA9I7e5Lnep9ovVC0x7BbyU9kF5Hp0mXAe9Q1jf+59K8b0lthkq6Kc3zntSlRMvzH66UtEjSnyS1SSKSPinpXwvjfy/p/6Xhm5R1prZEFTpUa72HI+kLki5JwwdI+mVq/1tJB6XyKWm7PSSpT3a5YLW3w/d9ZP3CW4D725n2P4HDgEOBYcB9hS+4CcBbImKVpONajU8n65LiSEmDgLsk3ca2nZxtBE6MiL9IGkfWjUkj2fMfvhARp0KWgAptvgL8ISLOkPRu4KoUH8BBZAluD2C5pG9HxGuFtj8F7ga+mMY/CMxMwx+LiD9L2i2t4/VR6LeoE7PJ7rpdIeltwCzg3cDFwEkRsVbSXlXOy7ZzTgrW3x0LXBsRW8k69LsDOBLYBCyOiFWFusXx9wCHSPpAGt+TrAfOPxbq7wz8h6TDgK1ANecwjgXeDxARt0t6Q+oPB+DnEbEZ2CxpI9nhrzUtDSOiOe1FHEXW/ceBwF1p8mckvS8Nj06xdpoUlPUwezRwXepGCGBQer8L+KGkeWT9IZk5KVi/sAT4QDvTKnVp3OLlDsYFfDoifrXNzLLnNrT4HLCBbC9kJ+AvVcTaURfLmwtlW6n8//cT4O+Ax4AbIyLSnsgJwNsj4hVJi4BdW7XbwraHg1um7wQ8HxGHtQkq4ty05/Be4EFJh3Vh78O2Uz6nYP3B7cAgSR9vKZB0pKR3kXU//MF0jmA48E5gcRXz/BXwSWVdcyPpzakHy6I9gfUR8Tfgo2SPZAR4kewQUCV3Ah9O8zwOeCYKXUBX4Qayp9CdSZYgWuJ4LiWEg8ge+9naBmBE2jMZBJwKeffTqyRNSTFJ6bnMkg6IiHsj4mLgGbbtGtp2UE4K1udF1mvj+4AT0yWpS8ie0byO7Kqkh8l6oLwd+GJEPF3FbL8PLAUeSCdov0vbX+6zgGmS7iE7dNSyp/EwsCWdoP1cqzaXAI2SHiY7IT2NLojsUZlLyR4u05LcfgkMTPO8lKzH0NbtXgO+SvZkulvI9jRafBg4W1JLL50tj5H8Rjo5/yhZMnuoK7Ha9sm9pJqZWc57CmZmlnNSMDOznJOCmZnlnBTMzCznpGBmZjknBTMzyzkpmJlZ7r8ABDpZDIChmGwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(non_sig['pearson'], bins=20, edgecolor='black')\n",
    "plt.xlabel('Correlation values')\n",
    "plt.ylabel('Frequency')\n",
    "plt.title('Non-associated variants')\n",
    "plt.savefig('non-sig_correlation.png', dpi = 300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c82f4e-570d-4d0c-b131-6f2ed51ad19f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fecb8e9-92c5-4c24-91e4-f675df161c23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1785c486-9c2e-4f3c-afac-129f3bcfe87e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
