{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a2b8b81d-48be-489b-a28f-3856205f52f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy as sp \n",
    "import numpy as np\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "313f2c5f-27a9-4c03-a99f-37e559e405dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain Identifier</th>\n",
       "      <th>Strain name</th>\n",
       "      <th>Evolution Experiment?</th>\n",
       "      <th>Parental Strain</th>\n",
       "      <th>Alternative Strain Name</th>\n",
       "      <th>Full Strain Name After Genome Analysis</th>\n",
       "      <th>ENA Accession</th>\n",
       "      <th>Alternative Genome Source</th>\n",
       "      <th>Reference (PMID)</th>\n",
       "      <th>Isolation origin</th>\n",
       "      <th>Year</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>German BSL level</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NT12447</td>\n",
       "      <td>B367</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KC329</td>\n",
       "      <td>Escherichia coli B367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Broad institute Antibiotic resistance collection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Australian magpie Coffs</td>\n",
       "      <td>2001</td>\n",
       "      <td>Commensal strain</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NT12619</td>\n",
       "      <td>S17</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli S17</td>\n",
       "      <td>PRJNA189395</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23516222</td>\n",
       "      <td>Broiler chick with septicemia</td>\n",
       "      <td>2013</td>\n",
       "      <td>Pathogenic strain</td>\n",
       "      <td>2.0</td>\n",
       "      <td>APEC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Strain Identifier Strain name  Evolution Experiment? Parental Strain  \\\n",
       "0           NT12447        B367                      0             NaN   \n",
       "1           NT12619         S17                      0             NaN   \n",
       "\n",
       "  Alternative Strain Name Full Strain Name After Genome Analysis  \\\n",
       "0                   KC329                  Escherichia coli B367   \n",
       "1                     NaN                   Escherichia coli S17   \n",
       "\n",
       "  ENA Accession                         Alternative Genome Source  \\\n",
       "0           NaN  Broad institute Antibiotic resistance collection   \n",
       "1   PRJNA189395                                               NaN   \n",
       "\n",
       "  Reference (PMID)               Isolation origin  Year          Phenotype  \\\n",
       "0              NaN       Australian magpie Coffs   2001   Commensal strain   \n",
       "1         23516222  Broiler chick with septicemia  2013  Pathogenic strain   \n",
       "\n",
       "   German BSL level Notes  \n",
       "0               1.0   NaN  \n",
       "1               2.0  APEC  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strains = pd.read_csv('mystrains.tsv', sep = '\\t')\n",
    "strains.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20337167-11f6-44bc-a441-3b3693a91999",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene-id</th>\n",
       "      <th>cds-NP_414543.1</th>\n",
       "      <th>cds-NP_414544.1</th>\n",
       "      <th>cds-NP_414545.1</th>\n",
       "      <th>cds-NP_414546.1</th>\n",
       "      <th>cds-NP_414547.1</th>\n",
       "      <th>cds-NP_414548.1</th>\n",
       "      <th>cds-NP_414549.1</th>\n",
       "      <th>cds-NP_414550.1</th>\n",
       "      <th>cds-NP_414551.1</th>\n",
       "      <th>cds-NP_414552.1</th>\n",
       "      <th>...</th>\n",
       "      <th>cds-NP_418811.2</th>\n",
       "      <th>cds-NP_418812.1</th>\n",
       "      <th>cds-NP_418813.1</th>\n",
       "      <th>cds-NP_418814.1</th>\n",
       "      <th>cds-NP_418815.1</th>\n",
       "      <th>cds-NP_418816.1</th>\n",
       "      <th>cds-NP_418817.1</th>\n",
       "      <th>cds-NP_418818.1</th>\n",
       "      <th>cds-NP_418819.1</th>\n",
       "      <th>cds-NP_418820.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NT12189</th>\n",
       "      <td>-1.396155</td>\n",
       "      <td>-1.086905</td>\n",
       "      <td>-1.036221</td>\n",
       "      <td>-1.127845</td>\n",
       "      <td>-0.170443</td>\n",
       "      <td>0.154180</td>\n",
       "      <td>0.743706</td>\n",
       "      <td>0.231421</td>\n",
       "      <td>-1.702605</td>\n",
       "      <td>2.666265</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.026376</td>\n",
       "      <td>0.058862</td>\n",
       "      <td>0.081481</td>\n",
       "      <td>0.324293</td>\n",
       "      <td>0.111350</td>\n",
       "      <td>0.357424</td>\n",
       "      <td>3.099915</td>\n",
       "      <td>0.239958</td>\n",
       "      <td>0.529843</td>\n",
       "      <td>0.288563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12177</th>\n",
       "      <td>-2.021794</td>\n",
       "      <td>-1.487570</td>\n",
       "      <td>-1.481607</td>\n",
       "      <td>-0.445211</td>\n",
       "      <td>0.236060</td>\n",
       "      <td>-0.659636</td>\n",
       "      <td>-0.199121</td>\n",
       "      <td>0.661353</td>\n",
       "      <td>-0.481898</td>\n",
       "      <td>0.949621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.682059</td>\n",
       "      <td>-0.349514</td>\n",
       "      <td>-0.762992</td>\n",
       "      <td>-0.336101</td>\n",
       "      <td>-0.088115</td>\n",
       "      <td>0.407631</td>\n",
       "      <td>0.086031</td>\n",
       "      <td>0.108566</td>\n",
       "      <td>0.537271</td>\n",
       "      <td>0.573327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12130</th>\n",
       "      <td>-0.956499</td>\n",
       "      <td>-0.586719</td>\n",
       "      <td>-0.499014</td>\n",
       "      <td>-1.345601</td>\n",
       "      <td>0.244311</td>\n",
       "      <td>-0.234788</td>\n",
       "      <td>0.440232</td>\n",
       "      <td>0.620264</td>\n",
       "      <td>-1.095199</td>\n",
       "      <td>0.796310</td>\n",
       "      <td>...</td>\n",
       "      <td>0.574220</td>\n",
       "      <td>0.061330</td>\n",
       "      <td>-0.116296</td>\n",
       "      <td>0.051911</td>\n",
       "      <td>0.330066</td>\n",
       "      <td>0.430785</td>\n",
       "      <td>-0.603411</td>\n",
       "      <td>0.244223</td>\n",
       "      <td>-0.138190</td>\n",
       "      <td>0.107360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 4201 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene-id  cds-NP_414543.1  cds-NP_414544.1  cds-NP_414545.1  cds-NP_414546.1  \\\n",
       "NT12189        -1.396155        -1.086905        -1.036221        -1.127845   \n",
       "NT12177        -2.021794        -1.487570        -1.481607        -0.445211   \n",
       "NT12130        -0.956499        -0.586719        -0.499014        -1.345601   \n",
       "\n",
       "gene-id  cds-NP_414547.1  cds-NP_414548.1  cds-NP_414549.1  cds-NP_414550.1  \\\n",
       "NT12189        -0.170443         0.154180         0.743706         0.231421   \n",
       "NT12177         0.236060        -0.659636        -0.199121         0.661353   \n",
       "NT12130         0.244311        -0.234788         0.440232         0.620264   \n",
       "\n",
       "gene-id  cds-NP_414551.1  cds-NP_414552.1  ...  cds-NP_418811.2  \\\n",
       "NT12189        -1.702605         2.666265  ...        -0.026376   \n",
       "NT12177        -0.481898         0.949621  ...        -0.682059   \n",
       "NT12130        -1.095199         0.796310  ...         0.574220   \n",
       "\n",
       "gene-id  cds-NP_418812.1  cds-NP_418813.1  cds-NP_418814.1  cds-NP_418815.1  \\\n",
       "NT12189         0.058862         0.081481         0.324293         0.111350   \n",
       "NT12177        -0.349514        -0.762992        -0.336101        -0.088115   \n",
       "NT12130         0.061330        -0.116296         0.051911         0.330066   \n",
       "\n",
       "gene-id  cds-NP_418816.1  cds-NP_418817.1  cds-NP_418818.1  cds-NP_418819.1  \\\n",
       "NT12189         0.357424         3.099915         0.239958         0.529843   \n",
       "NT12177         0.407631         0.086031         0.108566         0.537271   \n",
       "NT12130         0.430785        -0.603411         0.244223        -0.138190   \n",
       "\n",
       "gene-id  cds-NP_418820.1  \n",
       "NT12189         0.288563  \n",
       "NT12177         0.573327  \n",
       "NT12130         0.107360  \n",
       "\n",
       "[3 rows x 4201 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deg = pd.read_csv('deseqoutput_protein-ids.tsv', sep = '\\t', index_col = 0).T\n",
    "deg.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ee0136f0-4c5d-4c4a-8788-dd1a6887d6e9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>cds-NP_418030.1-793</th>\n",
       "      <th>cds-NP_417082.1-613</th>\n",
       "      <th>cds-NP_415117.2-501</th>\n",
       "      <th>cds-NP_416726.1-2615</th>\n",
       "      <th>cds-NP_417082.1-607</th>\n",
       "      <th>cds-NP_416801.2-1017</th>\n",
       "      <th>cds-NP_416770.1-4</th>\n",
       "      <th>cds-YP_026226.4-1448</th>\n",
       "      <th>cds-NP_416607.1-2059</th>\n",
       "      <th>...</th>\n",
       "      <th>cds-NP_415537.1-3644</th>\n",
       "      <th>cds-NP_416607.1-2068</th>\n",
       "      <th>cds-NP_416669.1-882</th>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "      <th>cds-NP_415708.1-3261</th>\n",
       "      <th>cds-NP_415592.1-4200</th>\n",
       "      <th>cds-NP_415129.1-35</th>\n",
       "      <th>cds-NP_415592.1-4214</th>\n",
       "      <th>cds-NP_416770.1-9</th>\n",
       "      <th>cds-NP_416834.1-4027</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NT12001</td>\n",
       "      <td>2461.616157</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1166.672266</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>2603.708627</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>1989.463318</td>\n",
       "      <td>...</td>\n",
       "      <td>1929.166732</td>\n",
       "      <td>1989.463318</td>\n",
       "      <td>1789.871250</td>\n",
       "      <td>2922.763810</td>\n",
       "      <td>2472.972260</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3780.695182</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>2472.846772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NT12002</td>\n",
       "      <td>2461.616157</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1166.672266</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>2603.708627</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>1989.463318</td>\n",
       "      <td>...</td>\n",
       "      <td>1929.166732</td>\n",
       "      <td>1989.463318</td>\n",
       "      <td>1789.871250</td>\n",
       "      <td>2922.763810</td>\n",
       "      <td>2472.972260</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3780.695182</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>2472.846772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NT12004</td>\n",
       "      <td>2461.616157</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1166.672266</td>\n",
       "      <td>13023.185487</td>\n",
       "      <td>2603.708627</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>1989.463318</td>\n",
       "      <td>...</td>\n",
       "      <td>1929.166732</td>\n",
       "      <td>1989.463318</td>\n",
       "      <td>1789.871250</td>\n",
       "      <td>2922.763810</td>\n",
       "      <td>2472.972260</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3780.695182</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>2472.846772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NT12012</td>\n",
       "      <td>2461.616157</td>\n",
       "      <td>3000.928338</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1166.672266</td>\n",
       "      <td>3000.928338</td>\n",
       "      <td>2651.387448</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>1677.181399</td>\n",
       "      <td>1989.463318</td>\n",
       "      <td>...</td>\n",
       "      <td>1929.166732</td>\n",
       "      <td>1989.463318</td>\n",
       "      <td>2078.272431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2472.972260</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3780.695182</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>2472.846772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NT12015</td>\n",
       "      <td>2461.616157</td>\n",
       "      <td>3132.884198</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3132.884198</td>\n",
       "      <td>2699.960566</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>1898.736786</td>\n",
       "      <td>...</td>\n",
       "      <td>1929.166732</td>\n",
       "      <td>1898.736786</td>\n",
       "      <td>2074.629602</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2472.972260</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3780.695182</td>\n",
       "      <td>1056.406243</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>2472.846772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>NT12460</td>\n",
       "      <td>2283.165598</td>\n",
       "      <td>4223.525815</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1166.592948</td>\n",
       "      <td>4223.525815</td>\n",
       "      <td>2743.978833</td>\n",
       "      <td>1613.648944</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>2806.113701</td>\n",
       "      <td>...</td>\n",
       "      <td>1846.500268</td>\n",
       "      <td>2806.113701</td>\n",
       "      <td>2078.272431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2512.953531</td>\n",
       "      <td>1171.587858</td>\n",
       "      <td>3441.374804</td>\n",
       "      <td>1171.587858</td>\n",
       "      <td>1613.648944</td>\n",
       "      <td>2472.846772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>NT12533</td>\n",
       "      <td>3771.499213</td>\n",
       "      <td>3684.816251</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1166.672266</td>\n",
       "      <td>3684.816251</td>\n",
       "      <td>2743.978833</td>\n",
       "      <td>1663.957144</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>2851.480967</td>\n",
       "      <td>...</td>\n",
       "      <td>1846.500268</td>\n",
       "      <td>2851.480967</td>\n",
       "      <td>2078.272431</td>\n",
       "      <td>3673.059652</td>\n",
       "      <td>2575.729013</td>\n",
       "      <td>1171.587858</td>\n",
       "      <td>3538.134647</td>\n",
       "      <td>1171.587858</td>\n",
       "      <td>1663.957144</td>\n",
       "      <td>2472.846772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>NT12552</td>\n",
       "      <td>3771.499213</td>\n",
       "      <td>4061.321551</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1166.672266</td>\n",
       "      <td>4061.321551</td>\n",
       "      <td>2699.960566</td>\n",
       "      <td>1663.957144</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>2851.480967</td>\n",
       "      <td>...</td>\n",
       "      <td>1846.500268</td>\n",
       "      <td>2851.480967</td>\n",
       "      <td>2078.272431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2512.953531</td>\n",
       "      <td>1171.587858</td>\n",
       "      <td>4338.663347</td>\n",
       "      <td>1171.587858</td>\n",
       "      <td>1663.957144</td>\n",
       "      <td>2472.846772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>NT12564</td>\n",
       "      <td>2610.289355</td>\n",
       "      <td>3547.091838</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1446.183168</td>\n",
       "      <td>3547.091838</td>\n",
       "      <td>2614.730877</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>2077.672606</td>\n",
       "      <td>...</td>\n",
       "      <td>2386.595383</td>\n",
       "      <td>2077.672606</td>\n",
       "      <td>2175.242841</td>\n",
       "      <td>4364.725987</td>\n",
       "      <td>2512.635765</td>\n",
       "      <td>1095.092977</td>\n",
       "      <td>3627.096143</td>\n",
       "      <td>1095.092977</td>\n",
       "      <td>3025.464313</td>\n",
       "      <td>4034.390926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>NT12616</td>\n",
       "      <td>3771.499213</td>\n",
       "      <td>3684.816251</td>\n",
       "      <td>12976.030829</td>\n",
       "      <td>1166.672266</td>\n",
       "      <td>3684.816251</td>\n",
       "      <td>2699.960566</td>\n",
       "      <td>1663.957144</td>\n",
       "      <td>1518.174637</td>\n",
       "      <td>2851.480967</td>\n",
       "      <td>...</td>\n",
       "      <td>1846.500268</td>\n",
       "      <td>2851.480967</td>\n",
       "      <td>2078.272431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2512.953531</td>\n",
       "      <td>1944.299225</td>\n",
       "      <td>3780.695182</td>\n",
       "      <td>1944.299225</td>\n",
       "      <td>1663.957144</td>\n",
       "      <td>2457.700744</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows × 4905 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      strain  cds-NP_418030.1-793  cds-NP_417082.1-613  cds-NP_415117.2-501  \\\n",
       "0    NT12001          2461.616157         13023.185487         12976.030829   \n",
       "1    NT12002          2461.616157         13023.185487         12976.030829   \n",
       "2    NT12004          2461.616157         13023.185487         12976.030829   \n",
       "3    NT12012          2461.616157          3000.928338         12976.030829   \n",
       "4    NT12015          2461.616157          3132.884198         12976.030829   \n",
       "..       ...                  ...                  ...                  ...   \n",
       "112  NT12460          2283.165598          4223.525815         12976.030829   \n",
       "113  NT12533          3771.499213          3684.816251         12976.030829   \n",
       "114  NT12552          3771.499213          4061.321551         12976.030829   \n",
       "115  NT12564          2610.289355          3547.091838         12976.030829   \n",
       "116  NT12616          3771.499213          3684.816251         12976.030829   \n",
       "\n",
       "     cds-NP_416726.1-2615  cds-NP_417082.1-607  cds-NP_416801.2-1017  \\\n",
       "0             1166.672266         13023.185487           2603.708627   \n",
       "1             1166.672266         13023.185487           2603.708627   \n",
       "2             1166.672266         13023.185487           2603.708627   \n",
       "3             1166.672266          3000.928338           2651.387448   \n",
       "4                     NaN          3132.884198           2699.960566   \n",
       "..                    ...                  ...                   ...   \n",
       "112           1166.592948          4223.525815           2743.978833   \n",
       "113           1166.672266          3684.816251           2743.978833   \n",
       "114           1166.672266          4061.321551           2699.960566   \n",
       "115           1446.183168          3547.091838           2614.730877   \n",
       "116           1166.672266          3684.816251           2699.960566   \n",
       "\n",
       "     cds-NP_416770.1-4  cds-YP_026226.4-1448  cds-NP_416607.1-2059  ...  \\\n",
       "0          3025.464313           1518.174637           1989.463318  ...   \n",
       "1          3025.464313           1518.174637           1989.463318  ...   \n",
       "2          3025.464313           1518.174637           1989.463318  ...   \n",
       "3          3025.464313           1677.181399           1989.463318  ...   \n",
       "4          3025.464313           1518.174637           1898.736786  ...   \n",
       "..                 ...                   ...                   ...  ...   \n",
       "112        1613.648944           1518.174637           2806.113701  ...   \n",
       "113        1663.957144           1518.174637           2851.480967  ...   \n",
       "114        1663.957144           1518.174637           2851.480967  ...   \n",
       "115        3025.464313           1518.174637           2077.672606  ...   \n",
       "116        1663.957144           1518.174637           2851.480967  ...   \n",
       "\n",
       "     cds-NP_415537.1-3644  cds-NP_416607.1-2068  cds-NP_416669.1-882  \\\n",
       "0             1929.166732           1989.463318          1789.871250   \n",
       "1             1929.166732           1989.463318          1789.871250   \n",
       "2             1929.166732           1989.463318          1789.871250   \n",
       "3             1929.166732           1989.463318          2078.272431   \n",
       "4             1929.166732           1898.736786          2074.629602   \n",
       "..                    ...                   ...                  ...   \n",
       "112           1846.500268           2806.113701          2078.272431   \n",
       "113           1846.500268           2851.480967          2078.272431   \n",
       "114           1846.500268           2851.480967          2078.272431   \n",
       "115           2386.595383           2077.672606          2175.242841   \n",
       "116           1846.500268           2851.480967          2078.272431   \n",
       "\n",
       "     cds-YP_588463.1-2320  cds-NP_415708.1-3261  cds-NP_415592.1-4200  \\\n",
       "0             2922.763810           2472.972260           1056.406243   \n",
       "1             2922.763810           2472.972260           1056.406243   \n",
       "2             2922.763810           2472.972260           1056.406243   \n",
       "3                     NaN           2472.972260           1056.406243   \n",
       "4                     NaN           2472.972260           1056.406243   \n",
       "..                    ...                   ...                   ...   \n",
       "112                   NaN           2512.953531           1171.587858   \n",
       "113           3673.059652           2575.729013           1171.587858   \n",
       "114                   NaN           2512.953531           1171.587858   \n",
       "115           4364.725987           2512.635765           1095.092977   \n",
       "116                   NaN           2512.953531           1944.299225   \n",
       "\n",
       "     cds-NP_415129.1-35  cds-NP_415592.1-4214  cds-NP_416770.1-9  \\\n",
       "0           3780.695182           1056.406243        3025.464313   \n",
       "1           3780.695182           1056.406243        3025.464313   \n",
       "2           3780.695182           1056.406243        3025.464313   \n",
       "3           3780.695182           1056.406243        3025.464313   \n",
       "4           3780.695182           1056.406243        3025.464313   \n",
       "..                  ...                   ...                ...   \n",
       "112         3441.374804           1171.587858        1613.648944   \n",
       "113         3538.134647           1171.587858        1663.957144   \n",
       "114         4338.663347           1171.587858        1663.957144   \n",
       "115         3627.096143           1095.092977        3025.464313   \n",
       "116         3780.695182           1944.299225        1663.957144   \n",
       "\n",
       "     cds-NP_416834.1-4027  \n",
       "0             2472.846772  \n",
       "1             2472.846772  \n",
       "2             2472.846772  \n",
       "3             2472.846772  \n",
       "4             2472.846772  \n",
       "..                    ...  \n",
       "112           2472.846772  \n",
       "113           2472.846772  \n",
       "114           2472.846772  \n",
       "115           4034.390926  \n",
       "116           2457.700744  \n",
       "\n",
       "[117 rows x 4905 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = 'tested_patterns/'\n",
    "full_table = None\n",
    "\n",
    "for gene in os.listdir(path):\n",
    "    if gene.startswith('cds'):\n",
    "        name = gene.split('-')[1]\n",
    "        name = 'cds-' + name\n",
    "        \n",
    "        #get hashed-patterns used in association analysis\n",
    "        association = pd.read_csv(path + gene, sep = '\\t', index_col = 0, names = ['variant', 'lrt-pvalue', 'beta', 'k-samples'])\n",
    "       \n",
    "        #get the representative kmer for the hased pattern\n",
    "        hashes = pd.read_csv('panfeed-renamed-fixed/' + name + '/kmers_to_hashes.tsv.gz', compression = 'gzip', sep = '\\t', index_col = 2)\n",
    "    \n",
    "        kmers = pd.read_csv('panfeed-renamed-fixed/' + name + '/kmers.tsv.gz', compression = 'gzip', sep = '\\t').set_index('k-mer')\n",
    "        kmers['strain'] = [k.split('_')[0] for k in kmers['strain']] # rename strains so they match those in the deg, and strain file for downstream analysis\n",
    "        \n",
    "        p = pd.merge(association, hashes, left_index = True, right_index = True, how = 'inner')\n",
    "        \n",
    "        p2 = kmers[kmers.index.isin(p['k-mer'])]\n",
    "        \n",
    "        #get predicted transcription rates for each gene kmer and strain\n",
    "        file = name + '.tsv.gz'\n",
    "        rates_files = 'tx-panfeed_230/'\n",
    "        if file in os.listdir(rates_files):\n",
    "            rates = pd.read_csv(rates_files + file , sep = '\\t', compression = 'gzip')\n",
    "            rates['strain'] = [r.split('_')[0] for r in rates.strain]\n",
    "            pr = pd.merge(p2, rates, on = 'strain', how = 'inner')\n",
    "\n",
    "            #select kmer that position is closest to the position of the predicted promoter and predicted tx_rate\n",
    "            pr['diff_start'] = pr['contig_start'] - pr['start']\n",
    "            pr['diff_end'] = pr['contig_end'] - pr['end']\n",
    "\n",
    "            #Group by 'strain' and deduce the minimum difference between kmers's start/end position and and that of the predicted promoter's start/end position, and then get the maximum predicted tx_rate\n",
    "            #this will return only tx_rates for strains that contain the gene kmer\n",
    "            tx = pr.groupby('strain').agg({ 'diff_start': 'min',\n",
    "                                           'diff_end': 'min',\n",
    "                                           'Tx_rate': 'max'}).reset_index()\n",
    "            tx = tx[['strain', 'Tx_rate']]\n",
    "\n",
    "            #get gene's predicted transcription rate for strains not present tx\n",
    "\n",
    "            tx2 = rates[~rates.strain.isin(tx.strain)]\n",
    "            tx2 = tx2.groupby('strain')['Tx_rate'].max().reset_index()\n",
    "            tx2 = tx2[['strain', 'Tx_rate']]\n",
    "\n",
    "            #merge the two dataframes to get one single dataframe\n",
    "            tx_r = pd.concat([tx, tx2]).rename(columns = {'Tx_rate' : gene.split('.tsv')[0]})\n",
    "\n",
    "            # if tx_r.empty:\n",
    "            #     continue\n",
    "\n",
    "            if full_table is None:\n",
    "                full_table = tx_r\n",
    "            else:\n",
    "                full_table = pd.merge(full_table, tx_r, on='strain', how='outer')\n",
    "full_table       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "55c247b5-35b0-48e6-8b5b-483b5865a0dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cds-NP_417082.1-613</th>\n",
       "      <th>cds-NP_417082.1-607</th>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <th>cds-NP_417420.1-2778</th>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <th>cds-NP_417420.1-2793</th>\n",
       "      <th>cds-NP_417420.1-2787</th>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <th>...</th>\n",
       "      <th>cds-NP_418440.1-1100</th>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <th>cds-NP_418618.1-1642</th>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <th>cds-NP_414578.2-87</th>\n",
       "      <th>cds-NP_414578.2-93</th>\n",
       "      <th>cds-NP_414578.2-78</th>\n",
       "      <th>cds-NP_417082.1-622</th>\n",
       "      <th>cds-NP_417082.1-636</th>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strain</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NT12001</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12002</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12004</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NT12012</th>\n",
       "      <td>-2.117602</td>\n",
       "      <td>-2.117602</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.318827</td>\n",
       "      <td>-1.464291</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.152345</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.960852</td>\n",
       "      <td>0.476826</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.117602</td>\n",
       "      <td>-2.117602</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 539 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         cds-NP_417082.1-613  cds-NP_417082.1-607  cds-NP_414578.2-75  \\\n",
       "strain                                                                  \n",
       "NT12001             0.000000             0.000000                 0.0   \n",
       "NT12002             0.000000             0.000000                 0.0   \n",
       "NT12004             0.000000             0.000000                 0.0   \n",
       "NT12012            -2.117602            -2.117602                 0.0   \n",
       "\n",
       "         cds-NP_418204.2-940  cds-NP_414880.2-2266  cds-NP_417420.1-2778  \\\n",
       "strain                                                                     \n",
       "NT12001             0.000000              0.000000              0.000000   \n",
       "NT12002             0.000000                   NaN              0.000000   \n",
       "NT12004             0.000000              0.000000              0.000000   \n",
       "NT12012             0.318827             -1.464291              0.000012   \n",
       "\n",
       "         cds-NP_418208.1-940  cds-NP_417420.1-2793  cds-NP_417420.1-2787  \\\n",
       "strain                                                                     \n",
       "NT12001                  0.0              0.000000              0.000000   \n",
       "NT12002                  0.0              0.000000              0.000000   \n",
       "NT12004                  0.0              0.000000              0.000000   \n",
       "NT12012                  0.0              0.000012              0.000012   \n",
       "\n",
       "         cds-NP_416773.1-1254  ...  cds-NP_418440.1-1100  \\\n",
       "strain                         ...                         \n",
       "NT12001                   0.0  ...              0.000000   \n",
       "NT12002                   0.0  ...              0.000000   \n",
       "NT12004                   0.0  ...              0.000000   \n",
       "NT12012                   0.0  ...              0.152345   \n",
       "\n",
       "         cds-NP_418554.1-1855  cds-NP_418618.1-1642  cds-NP_418686.1-718  \\\n",
       "strain                                                                     \n",
       "NT12001                   0.0              0.000000             0.000000   \n",
       "NT12002                   0.0              0.000000             0.000000   \n",
       "NT12004                   0.0              0.000000             0.000000   \n",
       "NT12012                   0.0             -0.960852             0.476826   \n",
       "\n",
       "         cds-NP_414578.2-87  cds-NP_414578.2-93  cds-NP_414578.2-78  \\\n",
       "strain                                                                \n",
       "NT12001                 0.0                 0.0                 0.0   \n",
       "NT12002                 0.0                 0.0                 0.0   \n",
       "NT12004                 0.0                 0.0                 0.0   \n",
       "NT12012                 0.0                 0.0                 0.0   \n",
       "\n",
       "         cds-NP_417082.1-622  cds-NP_417082.1-636  cds-YP_588463.1-2320  \n",
       "strain                                                                   \n",
       "NT12001             0.000000             0.000000                   0.0  \n",
       "NT12002             0.000000             0.000000                   0.0  \n",
       "NT12004             0.000000             0.000000                   0.0  \n",
       "NT12012            -2.117602            -2.117602                   NaN  \n",
       "\n",
       "[4 rows x 539 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#correlation between log2Foldchange and predicted transcription rates \n",
    "#tx_rates of each divided by that of k12 and then log2 transformed.\n",
    "tx_rates_filtered = full_table.copy()\n",
    "df_new = np.log2(tx_rates_filtered/tx_rates_filtered.loc['NT12001'])\n",
    "df_new.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd186a8d-2060-412b-a8e8-cd7e2d544b4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete. Check the 'corr-plots' directory for the generated plots.\n"
     ]
    }
   ],
   "source": [
    "import warnings \n",
    "\n",
    "pearson = {}\n",
    "spearman = {}\n",
    "regress = {}\n",
    "\n",
    "def safe_correlation(x, y):\n",
    "    if len(set(x)) == 1 or len(set(y)) == 1:\n",
    "        return 0, 1  # Return correlation of 0 with p-value of 1 for constant input\n",
    "    return sp.stats.pearsonr(x, y)\n",
    "\n",
    "def safe_regression(x, y):\n",
    "    if len(set(x)) == 1 or len(set(y)) == 1:\n",
    "        return [0, 0, 0, 1, 0]  # Return dummy values for constant input\n",
    "    return stats.linregress(x, y)\n",
    "\n",
    "for var in df_new.columns:\n",
    "    g_var = 'cds-' + var.split('-')[1]\n",
    "    var_df = f_df[f_df['gene_name'] == var]\n",
    "    samples = set(var_df['k-samples'].str.cat(sep=',').split(','))\n",
    "    \n",
    "    if g_var in deg.columns:\n",
    "        # Merge predicted transcription rate and actual expression\n",
    "        d = pd.merge(df_new[var], deg[g_var], left_index=True, right_index=True, how='inner').dropna()\n",
    "        \n",
    "        # Filter to include only strains with the variant\n",
    "        d_variant = d[d.index.isin(samples)]\n",
    "        \n",
    "        if d_variant.empty:\n",
    "            print(f\"No data points with variant for {var}. Skipping.\")\n",
    "            continue\n",
    "        \n",
    "        d_variant.columns = ['Predicted transcription rate', 'log2foldchange']\n",
    "        \n",
    "        # Calculate correlations and regression\n",
    "        with warnings.catch_warnings():\n",
    "            warnings.simplefilter(\"ignore\")\n",
    "            p, pp = safe_correlation(d_variant['Predicted transcription rate'], d_variant['log2foldchange'])\n",
    "            s, ps = safe_correlation(sp.stats.rankdata(d_variant['Predicted transcription rate']), \n",
    "                                     sp.stats.rankdata(d_variant['log2foldchange']))\n",
    "            slope, intercept, r_value, p_value, std_err = safe_regression(d_variant['Predicted transcription rate'], \n",
    "                                                                          d_variant['log2foldchange'])\n",
    "        \n",
    "        pearson[var] = round(p, 2), pp\n",
    "        spearman[var] = round(s, 2), ps\n",
    "        regress[var] = [round(r_value, 2), p_value]\n",
    "        \n",
    "        # Create scatter plot\n",
    "        plt.figure(figsize=(5, 4))\n",
    "        plt.rcParams.update({'font.size': 14})\n",
    "        sns.scatterplot(data=d_variant, x='Predicted transcription rate', y='log2foldchange')\n",
    "        \n",
    "        if len(set(d_variant['Predicted transcription rate'])) > 1 and len(set(d_variant['log2foldchange'])) > 1:\n",
    "            sns.regplot(data=d_variant, x='Predicted transcription rate', y='log2foldchange', \n",
    "                        scatter=False, \n",
    "                        label=f'r={p:.2f}')\n",
    "        else:\n",
    "            plt.plot([], [], color='red', label='Constant input')\n",
    "        \n",
    "        plt.ylabel( '$\\log_2$(FC)\\n(Actual Expression)')\n",
    "        plt.xlabel('Predicted transcription rate\\nrelative to $\\it{E. coli}$ K-12')\n",
    "        plt.title(f\"{g_var}\")\n",
    "        plt.tight_layout()\n",
    "        plt.legend()\n",
    "        plt.savefig(f'corr-plots_var_present/{var}_variant_only.svg', dpi=300)\n",
    "        plt.close()\n",
    "\n",
    "print(\"Analysis complete. Check the 'corr-plots' directory for the generated plots.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb6f3560-1c0f-4461-9c52-d52bfef1e1a5",
   "metadata": {},
   "source": [
    "Comparing predicted expression rates of strains with the variants vs those without the variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "162d2442-b352-4c38-9cdc-b53a6616a981",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(PearsonRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  warnings.warn(SpearmanRConstantInputWarning())\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:170: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  slope = ssxym / ssxm\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:187: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  slope_stderr = np.sqrt((1 - r**2) * ssym / ssxm / df)\n",
      "/usr/local/Cellar/jupyterlab/3.2.8/libexec/lib/python3.9/site-packages/scipy/stats/_stats_mstats_common.py:194: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  intercept_stderr = slope_stderr * np.sqrt(ssxm + xmean**2)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete. Check the 'corr-plots' directory for the generated plots.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "pearson = {}\n",
    "spearman = {}\n",
    "regress = {}\n",
    "\n",
    "for var in df_new.columns:\n",
    "    g_var = 'cds-' + var.split('-')[1]\n",
    "    var_df = f_df[f_df['gene_name'] == var]\n",
    "    samples = set(var_df['k-samples'].str.cat(sep=',').split(','))\n",
    "    \n",
    "    if g_var in deg.columns:\n",
    "        # Merge predicted transcription rate and actual expression\n",
    "        d = pd.merge(df_new[var], deg[g_var], left_index=True, right_index=True, how='inner').dropna()\n",
    "        \n",
    "        # Add a column to indicate variant presence\n",
    "        d['variant_present'] = d.index.isin(samples).astype(int)\n",
    "        d.columns = ['Predicted transcription rate', 'log2foldchange', 'variant_present']\n",
    "        \n",
    "        # Separate dataframes for variant present and absent\n",
    "        d_variant = d[d['variant_present'] == 1]\n",
    "        d_no_variant = d[d['variant_present'] == 0]\n",
    "        \n",
    "        # Calculate correlations and regression for both cases\n",
    "        for case, data in [(\"Variant\", d_variant), (\"No Variant\", d_no_variant)]:\n",
    "            if not data.empty:\n",
    "                p, pp = sp.stats.pearsonr(data['Predicted transcription rate'], data['log2foldchange'])\n",
    "                s, ps = sp.stats.spearmanr(data['Predicted transcription rate'], data['log2foldchange'])\n",
    "                slope, intercept, r_value, p_value, std_err = stats.linregress(data['Predicted transcription rate'], data['log2foldchange'])\n",
    "                \n",
    "                pearson[f\"{var}_{case}\"] = round(p, 2), pp\n",
    "                spearman[f\"{var}_{case}\"] = round(s, 2), ps\n",
    "                regress[f\"{var}_{case}\"] = [round(r_value, 2), p_value]\n",
    "        \n",
    "        # Create scatter plot\n",
    "        plt.figure(figsize=(10, 6))\n",
    "        sns.scatterplot(data=d, x='Predicted transcription rate', y='log2foldchange', \n",
    "                        hue='variant_present', palette={0: 'blue', 1: 'red'}, alpha=0.7)\n",
    "        \n",
    "        # Add regression lines for both cases\n",
    "        for case, data, color in [(\"Variant\", d_variant, 'red'), (\"No Variant\", d_no_variant, 'blue')]:\n",
    "            if not data.empty:\n",
    "                sns.regplot(data=data, x='Predicted transcription rate', y='log2foldchange', \n",
    "                            scatter=False, color=color, label=f'{case} (r={pearson[f\"{var}_{case}\"][0]:.2f}, p={pearson[f\"{var}_{case}\"][1]:.3f})')\n",
    "        \n",
    "        plt.ylabel('log2foldchange (Actual Expression)')\n",
    "        plt.xlabel('Predicted transcription rate')\n",
    "        plt.title(f\"{g_var}\")\n",
    "        plt.legend(title='Variant present')\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f'corr-plots/{var}_comparison.svg', dpi=300)\n",
    "        plt.close()\n",
    "\n",
    "print(\"Analysis complete. Check the 'corr-plots' directory for the generated plots.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3d039aa3-a162-444e-8323-c63893305ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save computed correlation values to dataframe for downstream analysis\n",
    "df_p = pd.DataFrame(pearson).T.rename(columns = {0: 'pearson', 1: 'pearson(p-value)'})\n",
    "df_sp = pd.DataFrame(spearman).T.rename(columns = {0:'spearman', 1:'spearman(p-value)'})\n",
    "linregress = pd.DataFrame(regress).T.rename(columns = {0: 'r_value', 1: 'regress(p-value)'})    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0506d6c2-ea59-4611-84ca-3d44092a954d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>association</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_418030.1-793</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417082.1-613</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415117.2-501</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416726.1-2615</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417082.1-607</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415592.1-4200</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415129.1-35</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_415592.1-4214</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416770.1-9</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416834.1-4027</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4904 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      association\n",
       "cds-NP_418030.1-793             0\n",
       "cds-NP_417082.1-613             0\n",
       "cds-NP_415117.2-501             0\n",
       "cds-NP_416726.1-2615            0\n",
       "cds-NP_417082.1-607             0\n",
       "...                           ...\n",
       "cds-NP_415592.1-4200            0\n",
       "cds-NP_415129.1-35              0\n",
       "cds-NP_415592.1-4214            0\n",
       "cds-NP_416770.1-9               0\n",
       "cds-NP_416834.1-4027            0\n",
       "\n",
       "[4904 rows x 1 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outcomes = {}\n",
    "for i in tx_rates.columns:\n",
    "    name = i + '.tsv'\n",
    "    a = pd.read_csv('/Users/bamudamaris/ownCloud2/model/ecoli/validation_2024/tested_patterns/' + name,  sep = '\\t', index_col = 0, names = ['pattern', 'lrt-pvalue', 'beta', 'k-samples'])\n",
    "    a['assigned'] = a['lrt-pvalue'].apply(lambda x: 1 if x <= 6.95E-06 else 0)\n",
    "    outcomes[i] = a['assigned'].values.tolist()\n",
    "outcome = pd.DataFrame(outcomes).T.rename(columns = {0: 'association'}) \n",
    "outcome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "eb0e07a8-3ed0-475f-b57b-bb2a10ff92f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.151193e-01</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9.478283e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.448540e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.237597e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9.302139e-01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.966887e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>5.351611e-01</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.387368e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>0.21</td>\n",
       "      <td>5.638126e-02</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>3.137891e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418085.2-2294</th>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>0.31</td>\n",
       "      <td>5.926000e-02</td>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.465319e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417610.1-1604</th>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4.008327e-21</td>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.889370e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>2.683375e-02</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>4.712333e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.280097e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.488176e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>6.809198e-03</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.856504e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pearson  pearson(p-value)  spearman  spearman(p-value)  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01      0.02       8.151193e-01   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01     -0.04       6.448540e-01   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01      0.01       9.302139e-01   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01     -0.06       5.351611e-01   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01      0.21       5.638126e-02   \n",
       "...                       ...               ...       ...                ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02      0.31       5.926000e-02   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18      0.74       4.008327e-21   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05     -0.21       2.683375e-02   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01     -0.04       7.280097e-01   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02     -0.39       6.809198e-03   \n",
       "\n",
       "                      r_value  regress(p-value)  association  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01            1   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01            1   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01            1   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01            1   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01            1   \n",
       "...                       ...               ...          ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02            1   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18            1   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05            1   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01            1   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02            1   \n",
       "\n",
       "                      adjusted_p_values  \n",
       "cds-NP_414578.2-75         9.478283e-01  \n",
       "cds-NP_418204.2-940        8.237597e-01  \n",
       "cds-NP_414880.2-2266       6.966887e-01  \n",
       "cds-NP_418208.1-940        6.387368e-01  \n",
       "cds-NP_416773.1-1254       3.137891e-01  \n",
       "...                                 ...  \n",
       "cds-NP_418085.2-2294       1.465319e-01  \n",
       "cds-NP_417610.1-1604       1.889370e-16  \n",
       "cds-NP_418554.1-1855       4.712333e-04  \n",
       "cds-NP_418686.1-718        8.488176e-01  \n",
       "cds-YP_588463.1-2320       1.856504e-01  \n",
       "\n",
       "[267 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#combine dataframes\n",
    "# prediction = pd.concat([df_p, df_sp, linregress, outcome], axis = 1).dropna()\n",
    "# prediction['association'] = prediction['association'].astype(int)\n",
    "prediction = pd.read_csv('asso-correlation-panfeed230-sig.tsv', sep = '\\t', index_col = 0)\n",
    "prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cab872bd-87b0-4647-a51f-13eb78d5e948",
   "metadata": {},
   "source": [
    "Focus on significantly associated variants only "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ef92125b-9fa4-40ee-813f-323ae90bff81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.151193e-01</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9.478283e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.448540e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.237597e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9.302139e-01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.966887e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>5.351611e-01</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.387368e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>0.21</td>\n",
       "      <td>5.638126e-02</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>3.137891e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418085.2-2294</th>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>0.31</td>\n",
       "      <td>5.926000e-02</td>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.465319e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417610.1-1604</th>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4.008327e-21</td>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.889370e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>2.683375e-02</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>4.712333e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.280097e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.488176e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>6.809198e-03</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.856504e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pearson  pearson(p-value)  spearman  spearman(p-value)  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01      0.02       8.151193e-01   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01     -0.04       6.448540e-01   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01      0.01       9.302139e-01   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01     -0.06       5.351611e-01   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01      0.21       5.638126e-02   \n",
       "...                       ...               ...       ...                ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02      0.31       5.926000e-02   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18      0.74       4.008327e-21   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05     -0.21       2.683375e-02   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01     -0.04       7.280097e-01   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02     -0.39       6.809198e-03   \n",
       "\n",
       "                      r_value  regress(p-value)  association  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01            1   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01            1   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01            1   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01            1   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01            1   \n",
       "...                       ...               ...          ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02            1   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18            1   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05            1   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01            1   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02            1   \n",
       "\n",
       "                      adjusted_p_values  \n",
       "cds-NP_414578.2-75         9.478283e-01  \n",
       "cds-NP_418204.2-940        8.237597e-01  \n",
       "cds-NP_414880.2-2266       6.966887e-01  \n",
       "cds-NP_418208.1-940        6.387368e-01  \n",
       "cds-NP_416773.1-1254       3.137891e-01  \n",
       "...                                 ...  \n",
       "cds-NP_418085.2-2294       1.465319e-01  \n",
       "cds-NP_417610.1-1604       1.889370e-16  \n",
       "cds-NP_418554.1-1855       4.712333e-04  \n",
       "cds-NP_418686.1-718        8.488176e-01  \n",
       "cds-YP_588463.1-2320       1.856504e-01  \n",
       "\n",
       "[267 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#perform multiple test corrections\n",
    "\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "sig = prediction[prediction['association'] == 1]\n",
    "\n",
    "\n",
    "# Perform multiple test correlation using Benjamini-Hochberg (FDR) correction\n",
    "sig['adjusted_p_values'] = multipletests(sig['pearson(p-value)'], method='fdr_bh')[1]\n",
    "\n",
    "#sig.to_csv('asso-correlation-panfeed230-sig.tsv', sep = '\\t')\n",
    "\n",
    "sig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5b0bb355-1738-4aa0-b627-8867f3afa0c3",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For threshold 0.001, number of significant correlations is: 22 (8.24%)\n",
      "For threshold 0.002, number of significant correlations is: 22 (8.24%)\n",
      "For threshold 0.003, number of significant correlations is: 25 (9.36%)\n",
      "For threshold 0.004, number of significant correlations is: 25 (9.36%)\n",
      "For threshold 0.005, number of significant correlations is: 25 (9.36%)\n",
      "For threshold 0.0001, number of significant correlations is: 21 (7.87%)\n",
      "For threshold 0.01, number of significant correlations is: 26 (9.74%)\n",
      "For threshold 0.05, number of significant correlations is: 34 (12.73%)\n"
     ]
    }
   ],
   "source": [
    "thresholds = [0.001,0.002, 0.003, 0.004, 0.005, 0.0001, 0.01, 0.05]  \n",
    "\n",
    "total_var = len(sig)\n",
    "\n",
    "for threshold in thresholds:\n",
    "    significant_correlations = sig[(sig['pearson'] > 0) & (sig['adjusted_p_values'] <= threshold)]\n",
    "    num_significant = len(significant_correlations)\n",
    "    percentage = (num_significant / total_var) * 100\n",
    "    print(f\"For threshold {threshold}, number of significant correlations is: {num_significant} ({percentage:.2f}%)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af44804-0ca5-4224-9073-c87b97d739e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.151193e-01</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>8.981294e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9.478283e-01</td>\n",
       "      <td>cds-NP_414578.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.448540e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>6.972647e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.237597e-01</td>\n",
       "      <td>cds-NP_418204.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9.302139e-01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>5.270828e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.966887e-01</td>\n",
       "      <td>cds-NP_414880.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418208.1-940</th>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>5.351611e-01</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>4.455788e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>6.387368e-01</td>\n",
       "      <td>cds-NP_418208.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_416773.1-1254</th>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>0.21</td>\n",
       "      <td>5.638126e-02</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.462598e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>3.137891e-01</td>\n",
       "      <td>cds-NP_416773.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418085.2-2294</th>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>0.31</td>\n",
       "      <td>5.926000e-02</td>\n",
       "      <td>0.33</td>\n",
       "      <td>4.280707e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.465319e-01</td>\n",
       "      <td>cds-NP_418085.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_417610.1-1604</th>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4.008327e-21</td>\n",
       "      <td>0.69</td>\n",
       "      <td>9.906809e-18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.889370e-16</td>\n",
       "      <td>cds-NP_417610.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418554.1-1855</th>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>2.683375e-02</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>6.530199e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>4.712333e-04</td>\n",
       "      <td>cds-NP_418554.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418686.1-718</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.280097e-01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>7.416543e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>8.488176e-01</td>\n",
       "      <td>cds-NP_418686.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-YP_588463.1-2320</th>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>6.809198e-03</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>6.049283e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>1.856504e-01</td>\n",
       "      <td>cds-YP_588463.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pearson  pearson(p-value)  spearman  spearman(p-value)  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01      0.02       8.151193e-01   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01     -0.04       6.448540e-01   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01      0.01       9.302139e-01   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01     -0.06       5.351611e-01   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01      0.21       5.638126e-02   \n",
       "...                       ...               ...       ...                ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02      0.31       5.926000e-02   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18      0.74       4.008327e-21   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05     -0.21       2.683375e-02   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01     -0.04       7.280097e-01   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02     -0.39       6.809198e-03   \n",
       "\n",
       "                      r_value  regress(p-value)  association  \\\n",
       "cds-NP_414578.2-75      -0.01      8.981294e-01            1   \n",
       "cds-NP_418204.2-940     -0.04      6.972647e-01            1   \n",
       "cds-NP_414880.2-2266     0.07      5.270828e-01            1   \n",
       "cds-NP_418208.1-940     -0.07      4.455788e-01            1   \n",
       "cds-NP_416773.1-1254     0.16      1.462598e-01            1   \n",
       "...                       ...               ...          ...   \n",
       "cds-NP_418085.2-2294     0.33      4.280707e-02            1   \n",
       "cds-NP_417610.1-1604     0.69      9.906809e-18            1   \n",
       "cds-NP_418554.1-1855    -0.37      6.530199e-05            1   \n",
       "cds-NP_418686.1-718     -0.04      7.416543e-01            1   \n",
       "cds-YP_588463.1-2320    -0.28      6.049283e-02            1   \n",
       "\n",
       "                      adjusted_p_values             gene  \n",
       "cds-NP_414578.2-75         9.478283e-01  cds-NP_414578.2  \n",
       "cds-NP_418204.2-940        8.237597e-01  cds-NP_418204.2  \n",
       "cds-NP_414880.2-2266       6.966887e-01  cds-NP_414880.2  \n",
       "cds-NP_418208.1-940        6.387368e-01  cds-NP_418208.1  \n",
       "cds-NP_416773.1-1254       3.137891e-01  cds-NP_416773.1  \n",
       "...                                 ...              ...  \n",
       "cds-NP_418085.2-2294       1.465319e-01  cds-NP_418085.2  \n",
       "cds-NP_417610.1-1604       1.889370e-16  cds-NP_417610.1  \n",
       "cds-NP_418554.1-1855       4.712333e-04  cds-NP_418554.1  \n",
       "cds-NP_418686.1-718        8.488176e-01  cds-NP_418686.1  \n",
       "cds-YP_588463.1-2320       1.856504e-01  cds-YP_588463.1  \n",
       "\n",
       "[267 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig['gene'] = ['cds-' + g.split('-')[1] for g in sig.index]\n",
    "sig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "911c3813-9d9f-4541-972c-ecb174d001f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pearson</th>\n",
       "      <th>pearson(p-value)</th>\n",
       "      <th>spearman</th>\n",
       "      <th>spearman(p-value)</th>\n",
       "      <th>r_value</th>\n",
       "      <th>regress(p-value)</th>\n",
       "      <th>association</th>\n",
       "      <th>adjusted_p_values</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-NP_414578.2-75</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.898129</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.815119</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.898129</td>\n",
       "      <td>1</td>\n",
       "      <td>0.947828</td>\n",
       "      <td>cds-NP_414578.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_418204.2-940</th>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.697265</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.644854</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.697265</td>\n",
       "      <td>1</td>\n",
       "      <td>0.823760</td>\n",
       "      <td>cds-NP_418204.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-NP_414880.2-2266</th>\n",
       "      <td>0.07</td>\n",
       "      <td>0.527083</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.930214</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.527083</td>\n",
       "      <td>1</td>\n",
       "      <td>0.696689</td>\n",
       "      <td>cds-NP_414880.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pearson  pearson(p-value)  spearman  spearman(p-value)  \\\n",
       "cds-NP_414578.2-75      -0.01          0.898129      0.02           0.815119   \n",
       "cds-NP_418204.2-940     -0.04          0.697265     -0.04           0.644854   \n",
       "cds-NP_414880.2-2266     0.07          0.527083      0.01           0.930214   \n",
       "\n",
       "                      r_value  regress(p-value)  association  \\\n",
       "cds-NP_414578.2-75      -0.01          0.898129            1   \n",
       "cds-NP_418204.2-940     -0.04          0.697265            1   \n",
       "cds-NP_414880.2-2266     0.07          0.527083            1   \n",
       "\n",
       "                      adjusted_p_values             gene  \n",
       "cds-NP_414578.2-75             0.947828  cds-NP_414578.2  \n",
       "cds-NP_418204.2-940            0.823760  cds-NP_418204.2  \n",
       "cds-NP_414880.2-2266           0.696689  cds-NP_414880.2  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = sig.copy()\n",
    "data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5d8fec99-3036-40a0-aad8-282f442bcf49",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = sig.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d0fa824-ef79-42a2-a944-151a43a2b49d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No correlation (r = 0.0): 5 (1.87%)\n",
      "Weak positive (0 < r ≤ 0.2): 83 (31.09%)\n",
      "Weak negative (-0.2 ≤ r < 0): 92 (34.46%)\n",
      "Moderate positive (0.2 < r < 0.8): 39 (14.61%)\n",
      "Moderate negative (-0.8 < r < -0.2): 41 (15.36%)\n",
      "Strong positive (r ≥ 0.8): 7 (2.62%)\n",
      "Strong negative (r ≤ -0.8): 0 (0.00%)\n",
      "\n",
      "Correlations > 0.2: 46 (17.23%)\n",
      "Correlations < -0.2: 41 (15.36%)\n"
     ]
    }
   ],
   "source": [
    "# Total number of correlations\n",
    "count = data.shape[0]\n",
    "\n",
    "# No correlation\n",
    "no_corr_df = data[data['pearson'] == 0.0]\n",
    "no_correlation = no_corr_df.shape[0]\n",
    "\n",
    "# Weak positive (0 < r ≤ 0.2)\n",
    "weak_pos_df = data[(data['pearson'] > 0.0) & (data['pearson'] <= 0.2)]\n",
    "weak_pos = weak_pos_df.shape[0]\n",
    "\n",
    "# Weak negative (-0.2 ≤ r < 0)\n",
    "weak_neg_df = data[(data['pearson'] < 0.0) & (data['pearson'] >= -0.2)]\n",
    "weak_neg = weak_neg_df.shape[0]\n",
    "\n",
    "# Moderate positive (0.2 < r < 0.8)\n",
    "moderate_pos_df = data[(data['pearson'] > 0.2) & (data['pearson'] < 0.8)]\n",
    "moderate_pos = moderate_pos_df.shape[0]\n",
    "\n",
    "# Moderate negative (-0.8 < r < -0.2)\n",
    "moderate_neg_df = data[(data['pearson'] < -0.2) & (data['pearson'] > -0.8)]\n",
    "moderate_neg = moderate_neg_df.shape[0]\n",
    "\n",
    "# Strong positive (r ≥ 0.8)\n",
    "strong_pos_df = data[data['pearson'] >= 0.8]\n",
    "strong_pos = strong_pos_df.shape[0]\n",
    "\n",
    "# Strong negative (r ≤ -0.8)\n",
    "strong_neg_df = data[data['pearson'] <= -0.8]\n",
    "strong_neg = strong_neg_df.shape[0]\n",
    "\n",
    "# Print summary with direction\n",
    "print(f\"No correlation (r = 0.0): {no_correlation} ({no_correlation/count*100:.2f}%)\")\n",
    "print(f\"Weak positive (0 < r ≤ 0.2): {weak_pos} ({weak_pos/count*100:.2f}%)\")\n",
    "print(f\"Weak negative (-0.2 ≤ r < 0): {weak_neg} ({weak_neg/count*100:.2f}%)\")\n",
    "print(f\"Moderate positive (0.2 < r < 0.8): {moderate_pos} ({moderate_pos/count*100:.2f}%)\")\n",
    "print(f\"Moderate negative (-0.8 < r < -0.2): {moderate_neg} ({moderate_neg/count*100:.2f}%)\")\n",
    "print(f\"Strong positive (r ≥ 0.8): {strong_pos} ({strong_pos/count*100:.2f}%)\")\n",
    "print(f\"Strong negative (r ≤ -0.8): {strong_neg} ({strong_neg/count*100:.2f}%)\")\n",
    "\n",
    "# Supervisor’s key interest: total counts >0.2 and < -0.2\n",
    "pos_corr = data[data['pearson'] > 0.2].shape[0]\n",
    "neg_corr = data[data['pearson'] < -0.2].shape[0]\n",
    "\n",
    "print(f\"\\nCorrelations > 0.2: {pos_corr} ({pos_corr/count*100:.2f}%)\")\n",
    "print(f\"Correlations < -0.2: {neg_corr} ({neg_corr/count*100:.2f}%)\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
